{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "changed-lawyer",
   "metadata": {},
   "source": [
    "# Set up"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "weekly-bronze",
   "metadata": {},
   "outputs": [],
   "source": [
    "# use with bioconda kernel"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "terminal-translator",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import urllib.request\n",
    "import os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "dying-cornwall",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/jovyan/data\n"
     ]
    }
   ],
   "source": [
    "! cd /home/jovyan/data/\n",
    "! pwd"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "mathematical-chair",
   "metadata": {},
   "source": [
    "# Import a list of all human TFs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "sustainable-accounting",
   "metadata": {},
   "outputs": [],
   "source": [
    "tf_df=pd.read_csv('http://humantfs.ccbr.utoronto.ca/download/v_1.01/DatabaseExtract_v_1.01.csv',index_col='HGNC symbol') # import a df of all human TFs\n",
    "tf_df=tf_df.drop(['Unnamed: 0'],axis=1) # tidy\n",
    "tf_list=tf_df.index.tolist() # convert gene symbol to list"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "neural-cookie",
   "metadata": {},
   "source": [
    "# Define a list of phenotypes and search terms for those"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "reserved-bridal",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Using the list of phenotypes (see notebook pt1), make a dictionary grouping them. Have remoced partial strings in case these are inappropraitely matched\n",
    "phenotype_dict={\n",
    "    'Heterotaxy':['Heterotaxy','Dextrocardia'],\n",
    "    'ASD':['ASD','Atrial septal defect','Patent foramen ovale'],\n",
    "    'VSD':['VSD','Ventricular septal defect'],\n",
    "    'AVSD':['AVSD','Atrioventricular septal defect'],\n",
    "    'Malformation of outflow tracts':['Truncus arteriosus','Malformation of outflow tracts','Fallot','fallot','Double outlet right ventricle','Pentalogy of fallot','Transposition of the great arteries','Transposition of the great vessels',],\n",
    "    'Functional single ventricle':['Functional single ventricle','Double-inlet left ventricle','Hypoplastic left heart syndrome','Hypoplastic right ventricle','Hypoplastic rv'],\n",
    "#    'Obstructive lesions':['Obstructive lesions','alv','stenosis'], # removed because I don't know how to define this term\n",
    "    'Cardiomyopathy':['Cardiomyopathy'],\n",
    "    'DCM':['DCM','Dilated cardiomyopathy'],\n",
    "    'HCM':['HCM','Hypertrophic cardiomyopathy'],\n",
    "    'LVNC':['LVNC','Left ventricular noncompaction'],\n",
    "    'Venous anomaly':['Total anomalous pulmonary venous return','Pulmonary vein stenosis','Partial anomalous pulmonary venous return','Persistent left superior vena cava'],\n",
    "    'Pulmonary artery':['Dilatation of the pulmonary trunk','Hypoplastic left pulmonary artery','Pulmonary artery dysplasia and hypoplasia','Pulmonary atresia','Pulmonary trunk and pulmonary artery absence'],\n",
    "    'Aortic arch':['Right aortic arch','Patent ductus arteriosus','Ductus arteriosus','Coarctation of the aorta','Interrupted aortic arch','Lsvc and left pulmonary artery arising from the ductus arteriosus'],\n",
    "    'Aorta':['Aberrant supraclavicular artery','Anomalous right subclavian artery','Aortic aneurysm','Aortic hypoplasia'],\n",
    "    'Coronary':['Anomalous left coronary artery from the pulmonary artery','Coronary artery anomaly'],\n",
    "    'VA valves':['Pulmonic stenosis','Absent pulmonary valve','Bicuspid aortic valve', 'Aortic stenosis','Dysplastic pulmonary valve','Pulmonary stenosis'],\n",
    "    'AV valves':['Dysplastic mitral valve','Dysplastic tricuspid valve',\"Ebstein's anomaly\",'Mitral atresia','Mitral stenosis','Mitral valve dysplasia','Tricuspid atresia','Tricuspid stenosis','Tricuspid valve atresia','Tricuspid valve dysplasia']\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "developmental-producer",
   "metadata": {},
   "source": [
    "# Download the (weekly updating) ClinVar database from the FTP server"
   ]
  },
  {
   "cell_type": "raw",
   "id": "surprising-strike",
   "metadata": {},
   "source": [
    "urllib.request.urlretrieve('https://ftp.ncbi.nlm.nih.gov/pub/clinvar/tab_delimited/variant_summary.txt.gz', '/home/jovyan/data/variant_summary.txt.gz')"
   ]
  },
  {
   "cell_type": "raw",
   "id": "vietnamese-lodging",
   "metadata": {},
   "source": [
    "! gunzip /home/jovyan/data/variant_summary.txt.gz"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "frequent-center",
   "metadata": {},
   "source": [
    "# Read in the database"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "considered-computer",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_5443/3889004697.py:1: DtypeWarning: Columns (18) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  clinvar_summary_df=pd.read_csv('/home/jovyan/data/variant_summary.txt',sep='\\t')\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "(3093519, 34)"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clinvar_summary_df=pd.read_csv('/home/jovyan/data/variant_summary.txt',sep='\\t')\n",
    "clinvar_summary_df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "noted-cabinet",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['#AlleleID', 'Type', 'Name', 'GeneID', 'GeneSymbol', 'HGNC_ID',\n",
       "       'ClinicalSignificance', 'ClinSigSimple', 'LastEvaluated', 'RS# (dbSNP)',\n",
       "       'nsv/esv (dbVar)', 'RCVaccession', 'PhenotypeIDS', 'PhenotypeList',\n",
       "       'Origin', 'OriginSimple', 'Assembly', 'ChromosomeAccession',\n",
       "       'Chromosome', 'Start', 'Stop', 'ReferenceAllele', 'AlternateAllele',\n",
       "       'Cytogenetic', 'ReviewStatus', 'NumberSubmitters', 'Guidelines',\n",
       "       'TestedInGTR', 'OtherIDs', 'SubmitterCategories', 'VariationID',\n",
       "       'PositionVCF', 'ReferenceAlleleVCF', 'AlternateAlleleVCF'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clinvar_summary_df.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "funky-prime",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#AlleleID</th>\n",
       "      <th>Type</th>\n",
       "      <th>Name</th>\n",
       "      <th>GeneID</th>\n",
       "      <th>GeneSymbol</th>\n",
       "      <th>HGNC_ID</th>\n",
       "      <th>ClinicalSignificance</th>\n",
       "      <th>ClinSigSimple</th>\n",
       "      <th>LastEvaluated</th>\n",
       "      <th>RS# (dbSNP)</th>\n",
       "      <th>...</th>\n",
       "      <th>ReviewStatus</th>\n",
       "      <th>NumberSubmitters</th>\n",
       "      <th>Guidelines</th>\n",
       "      <th>TestedInGTR</th>\n",
       "      <th>OtherIDs</th>\n",
       "      <th>SubmitterCategories</th>\n",
       "      <th>VariationID</th>\n",
       "      <th>PositionVCF</th>\n",
       "      <th>ReferenceAlleleVCF</th>\n",
       "      <th>AlternateAlleleVCF</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>15041</td>\n",
       "      <td>Indel</td>\n",
       "      <td>NM_014855.3(AP5Z1):c.80_83delinsTGCTGTAAACTGTA...</td>\n",
       "      <td>9907</td>\n",
       "      <td>AP5Z1</td>\n",
       "      <td>HGNC:22197</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>397704705</td>\n",
       "      <td>...</td>\n",
       "      <td>criteria provided, single submitter</td>\n",
       "      <td>2</td>\n",
       "      <td>-</td>\n",
       "      <td>N</td>\n",
       "      <td>ClinGen:CA215070,OMIM:613653.0001</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>4820844</td>\n",
       "      <td>GGAT</td>\n",
       "      <td>TGCTGTAAACTGTAACTGTAAA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>15041</td>\n",
       "      <td>Indel</td>\n",
       "      <td>NM_014855.3(AP5Z1):c.80_83delinsTGCTGTAAACTGTA...</td>\n",
       "      <td>9907</td>\n",
       "      <td>AP5Z1</td>\n",
       "      <td>HGNC:22197</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>397704705</td>\n",
       "      <td>...</td>\n",
       "      <td>criteria provided, single submitter</td>\n",
       "      <td>2</td>\n",
       "      <td>-</td>\n",
       "      <td>N</td>\n",
       "      <td>ClinGen:CA215070,OMIM:613653.0001</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>4781213</td>\n",
       "      <td>GGAT</td>\n",
       "      <td>TGCTGTAAACTGTAACTGTAAA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>15042</td>\n",
       "      <td>Deletion</td>\n",
       "      <td>NM_014855.3(AP5Z1):c.1413_1426del (p.Leu473fs)</td>\n",
       "      <td>9907</td>\n",
       "      <td>AP5Z1</td>\n",
       "      <td>HGNC:22197</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>1</td>\n",
       "      <td>Jun 29, 2010</td>\n",
       "      <td>397704709</td>\n",
       "      <td>...</td>\n",
       "      <td>no assertion criteria provided</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>N</td>\n",
       "      <td>ClinGen:CA215072,OMIM:613653.0002</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>4827360</td>\n",
       "      <td>GCTGCTGGACCTGCC</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 34 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   #AlleleID      Type                                               Name  \\\n",
       "0      15041     Indel  NM_014855.3(AP5Z1):c.80_83delinsTGCTGTAAACTGTA...   \n",
       "1      15041     Indel  NM_014855.3(AP5Z1):c.80_83delinsTGCTGTAAACTGTA...   \n",
       "2      15042  Deletion     NM_014855.3(AP5Z1):c.1413_1426del (p.Leu473fs)   \n",
       "\n",
       "   GeneID GeneSymbol     HGNC_ID ClinicalSignificance  ClinSigSimple  \\\n",
       "0    9907      AP5Z1  HGNC:22197           Pathogenic              1   \n",
       "1    9907      AP5Z1  HGNC:22197           Pathogenic              1   \n",
       "2    9907      AP5Z1  HGNC:22197           Pathogenic              1   \n",
       "\n",
       "  LastEvaluated  RS# (dbSNP)  ...                         ReviewStatus  \\\n",
       "0             -    397704705  ...  criteria provided, single submitter   \n",
       "1             -    397704705  ...  criteria provided, single submitter   \n",
       "2  Jun 29, 2010    397704709  ...       no assertion criteria provided   \n",
       "\n",
       "  NumberSubmitters Guidelines TestedInGTR                           OtherIDs  \\\n",
       "0                2          -           N  ClinGen:CA215070,OMIM:613653.0001   \n",
       "1                2          -           N  ClinGen:CA215070,OMIM:613653.0001   \n",
       "2                1          -           N  ClinGen:CA215072,OMIM:613653.0002   \n",
       "\n",
       "  SubmitterCategories VariationID PositionVCF ReferenceAlleleVCF  \\\n",
       "0                   3           2     4820844               GGAT   \n",
       "1                   3           2     4781213               GGAT   \n",
       "2                   1           3     4827360    GCTGCTGGACCTGCC   \n",
       "\n",
       "       AlternateAlleleVCF  \n",
       "0  TGCTGTAAACTGTAACTGTAAA  \n",
       "1  TGCTGTAAACTGTAACTGTAAA  \n",
       "2                       G  \n",
       "\n",
       "[3 rows x 34 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clinvar_summary_df.head(3)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "accredited-division",
   "metadata": {},
   "source": [
    "# Filter the database"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "presidential-edinburgh",
   "metadata": {},
   "source": [
    "- Pathogenic or Likely Pathogenic\n",
    "- Review status >1 submitter or expert\n",
    "- Last reviewed since 2016"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "alleged-niagara",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['#AlleleID', 'Type', 'Name', 'GeneID', 'GeneSymbol', 'HGNC_ID',\n",
       "       'ClinicalSignificance', 'ClinSigSimple', 'LastEvaluated', 'RS# (dbSNP)',\n",
       "       'nsv/esv (dbVar)', 'RCVaccession', 'PhenotypeIDS', 'PhenotypeList',\n",
       "       'Origin', 'OriginSimple', 'Assembly', 'ChromosomeAccession',\n",
       "       'Chromosome', 'Start', 'Stop', 'ReferenceAllele', 'AlternateAllele',\n",
       "       'Cytogenetic', 'ReviewStatus', 'NumberSubmitters', 'Guidelines',\n",
       "       'TestedInGTR', 'OtherIDs', 'SubmitterCategories', 'VariationID',\n",
       "       'PositionVCF', 'ReferenceAlleleVCF', 'AlternateAlleleVCF'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clinvar_summary_df.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "working-philadelphia",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Uncertain significance                                                1226639\n",
       "Likely benign                                                          828566\n",
       "Benign                                                                 402434\n",
       "Pathogenic                                                             256411\n",
       "Conflicting interpretations of pathogenicity                           133465\n",
       "                                                                       ...   \n",
       "Conflicting interpretations of pathogenicity; drug response; other          2\n",
       "Pathogenic; association; protective                                         2\n",
       "Conflicting interpretations of pathogenicity; Affects                       2\n",
       "Benign; association; confers sensitivity                                    2\n",
       "Conflicting interpretations of pathogenicity; protective                    2\n",
       "Name: ClinicalSignificance, Length: 80, dtype: int64"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clinvar_summary_df['ClinicalSignificance'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "junior-giant",
   "metadata": {},
   "outputs": [],
   "source": [
    "ClinicalSignificance=['Pathogenic',\n",
    "                      'Likely pathogenic', 'Pathogenic/Likely pathogenic'\n",
    "                     ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "divine-despite",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(406244, 34)"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clinvar_summary_df=clinvar_summary_df[clinvar_summary_df['ClinicalSignificance'].isin(ClinicalSignificance)]\n",
    "clinvar_summary_df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "strong-clarity",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "criteria provided, single submitter                     243267\n",
       "no assertion criteria provided                           82163\n",
       "criteria provided, multiple submitters, no conflicts     63520\n",
       "reviewed by expert panel                                 17244\n",
       "practice guideline                                          50\n",
       "Name: ReviewStatus, dtype: int64"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clinvar_summary_df['ReviewStatus'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "suited-sodium",
   "metadata": {},
   "outputs": [],
   "source": [
    "ReviewStatus=['criteria provided, multiple submitters, no conflicts','reviewed by expert panel']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "healthy-target",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(80764, 34)"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clinvar_summary_df=clinvar_summary_df[clinvar_summary_df['ReviewStatus'].isin(ReviewStatus)]\n",
    "clinvar_summary_df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "double-colonial",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2021    35318\n",
       "2022    18355\n",
       "2016     8366\n",
       "2020     5608\n",
       "2019     4103\n",
       "2017     4039\n",
       "2018     2357\n",
       "2013     2220\n",
       "2015      237\n",
       "2014       68\n",
       "2011       63\n",
       "2012        6\n",
       "2009        2\n",
       "Name: LastEvaluated_year, dtype: int64"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clinvar_summary_df['LastEvaluated_year']=clinvar_summary_df['LastEvaluated'].str.split(', ',expand=True)[1]\n",
    "clinvar_summary_df['LastEvaluated_year'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "coral-wholesale",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(78146, 35)"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clinvar_summary_df=clinvar_summary_df[clinvar_summary_df['LastEvaluated_year'].isin(['2022','2021','2020','2019','2018','2017','2016'])]\n",
    "clinvar_summary_df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "mysterious-inspector",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>PhenotypeIDS</th>\n",
       "      <th>PhenotypeList</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>MONDO:MONDO:0032624,MedGen:C4748791,OMIM:61824...</td>\n",
       "      <td>Mitochondrial complex 1 deficiency, nuclear ty...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>MONDO:MONDO:0032624,MedGen:C4748791,OMIM:61824...</td>\n",
       "      <td>Mitochondrial complex 1 deficiency, nuclear ty...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>MONDO:MONDO:0012984,MedGen:C2675204,OMIM:61267...</td>\n",
       "      <td>PHARC syndrome|not provided</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                         PhenotypeIDS  \\\n",
       "6   MONDO:MONDO:0032624,MedGen:C4748791,OMIM:61824...   \n",
       "7   MONDO:MONDO:0032624,MedGen:C4748791,OMIM:61824...   \n",
       "48  MONDO:MONDO:0012984,MedGen:C2675204,OMIM:61267...   \n",
       "\n",
       "                                        PhenotypeList  \n",
       "6   Mitochondrial complex 1 deficiency, nuclear ty...  \n",
       "7   Mitochondrial complex 1 deficiency, nuclear ty...  \n",
       "48                        PHARC syndrome|not provided  "
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clinvar_summary_df[['PhenotypeIDS', 'PhenotypeList']].head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "blind-austria",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_5443/2004218440.py:6: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  clinvar_summary_df['phenotype_'+str(list(phenotype_dict.keys())[phenotype])][clinvar_summary_df[phenotype_column].str.contains(case=False,pat='|'.join(list(phenotype_dict.values())[phenotype]))]=1\n",
      "/tmp/ipykernel_5443/2004218440.py:6: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  clinvar_summary_df['phenotype_'+str(list(phenotype_dict.keys())[phenotype])][clinvar_summary_df[phenotype_column].str.contains(case=False,pat='|'.join(list(phenotype_dict.values())[phenotype]))]=1\n",
      "/tmp/ipykernel_5443/2004218440.py:6: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  clinvar_summary_df['phenotype_'+str(list(phenotype_dict.keys())[phenotype])][clinvar_summary_df[phenotype_column].str.contains(case=False,pat='|'.join(list(phenotype_dict.values())[phenotype]))]=1\n",
      "/tmp/ipykernel_5443/2004218440.py:6: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  clinvar_summary_df['phenotype_'+str(list(phenotype_dict.keys())[phenotype])][clinvar_summary_df[phenotype_column].str.contains(case=False,pat='|'.join(list(phenotype_dict.values())[phenotype]))]=1\n",
      "/tmp/ipykernel_5443/2004218440.py:6: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  clinvar_summary_df['phenotype_'+str(list(phenotype_dict.keys())[phenotype])][clinvar_summary_df[phenotype_column].str.contains(case=False,pat='|'.join(list(phenotype_dict.values())[phenotype]))]=1\n",
      "/tmp/ipykernel_5443/2004218440.py:6: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  clinvar_summary_df['phenotype_'+str(list(phenotype_dict.keys())[phenotype])][clinvar_summary_df[phenotype_column].str.contains(case=False,pat='|'.join(list(phenotype_dict.values())[phenotype]))]=1\n",
      "/tmp/ipykernel_5443/2004218440.py:6: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  clinvar_summary_df['phenotype_'+str(list(phenotype_dict.keys())[phenotype])][clinvar_summary_df[phenotype_column].str.contains(case=False,pat='|'.join(list(phenotype_dict.values())[phenotype]))]=1\n",
      "/tmp/ipykernel_5443/2004218440.py:6: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  clinvar_summary_df['phenotype_'+str(list(phenotype_dict.keys())[phenotype])][clinvar_summary_df[phenotype_column].str.contains(case=False,pat='|'.join(list(phenotype_dict.values())[phenotype]))]=1\n",
      "/tmp/ipykernel_5443/2004218440.py:6: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  clinvar_summary_df['phenotype_'+str(list(phenotype_dict.keys())[phenotype])][clinvar_summary_df[phenotype_column].str.contains(case=False,pat='|'.join(list(phenotype_dict.values())[phenotype]))]=1\n",
      "/tmp/ipykernel_5443/2004218440.py:6: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  clinvar_summary_df['phenotype_'+str(list(phenotype_dict.keys())[phenotype])][clinvar_summary_df[phenotype_column].str.contains(case=False,pat='|'.join(list(phenotype_dict.values())[phenotype]))]=1\n",
      "/tmp/ipykernel_5443/2004218440.py:6: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  clinvar_summary_df['phenotype_'+str(list(phenotype_dict.keys())[phenotype])][clinvar_summary_df[phenotype_column].str.contains(case=False,pat='|'.join(list(phenotype_dict.values())[phenotype]))]=1\n",
      "/tmp/ipykernel_5443/2004218440.py:6: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  clinvar_summary_df['phenotype_'+str(list(phenotype_dict.keys())[phenotype])][clinvar_summary_df[phenotype_column].str.contains(case=False,pat='|'.join(list(phenotype_dict.values())[phenotype]))]=1\n",
      "/tmp/ipykernel_5443/2004218440.py:6: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  clinvar_summary_df['phenotype_'+str(list(phenotype_dict.keys())[phenotype])][clinvar_summary_df[phenotype_column].str.contains(case=False,pat='|'.join(list(phenotype_dict.values())[phenotype]))]=1\n",
      "/tmp/ipykernel_5443/2004218440.py:6: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  clinvar_summary_df['phenotype_'+str(list(phenotype_dict.keys())[phenotype])][clinvar_summary_df[phenotype_column].str.contains(case=False,pat='|'.join(list(phenotype_dict.values())[phenotype]))]=1\n",
      "/tmp/ipykernel_5443/2004218440.py:6: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  clinvar_summary_df['phenotype_'+str(list(phenotype_dict.keys())[phenotype])][clinvar_summary_df[phenotype_column].str.contains(case=False,pat='|'.join(list(phenotype_dict.values())[phenotype]))]=1\n",
      "/tmp/ipykernel_5443/2004218440.py:6: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  clinvar_summary_df['phenotype_'+str(list(phenotype_dict.keys())[phenotype])][clinvar_summary_df[phenotype_column].str.contains(case=False,pat='|'.join(list(phenotype_dict.values())[phenotype]))]=1\n",
      "/tmp/ipykernel_5443/2004218440.py:6: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  clinvar_summary_df['phenotype_'+str(list(phenotype_dict.keys())[phenotype])][clinvar_summary_df[phenotype_column].str.contains(case=False,pat='|'.join(list(phenotype_dict.values())[phenotype]))]=1\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#AlleleID</th>\n",
       "      <th>Type</th>\n",
       "      <th>Name</th>\n",
       "      <th>GeneID</th>\n",
       "      <th>GeneSymbol</th>\n",
       "      <th>HGNC_ID</th>\n",
       "      <th>ClinicalSignificance</th>\n",
       "      <th>ClinSigSimple</th>\n",
       "      <th>LastEvaluated</th>\n",
       "      <th>RS# (dbSNP)</th>\n",
       "      <th>...</th>\n",
       "      <th>phenotype_DCM</th>\n",
       "      <th>phenotype_HCM</th>\n",
       "      <th>phenotype_LVNC</th>\n",
       "      <th>phenotype_Venous anomaly</th>\n",
       "      <th>phenotype_Pulmonary artery</th>\n",
       "      <th>phenotype_Aortic arch</th>\n",
       "      <th>phenotype_Aorta</th>\n",
       "      <th>phenotype_Coronary</th>\n",
       "      <th>phenotype_VA valves</th>\n",
       "      <th>phenotype_AV valves</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>15044</td>\n",
       "      <td>single nucleotide variant</td>\n",
       "      <td>NM_017547.4(FOXRED1):c.694C&gt;T (p.Gln232Ter)</td>\n",
       "      <td>55572</td>\n",
       "      <td>FOXRED1</td>\n",
       "      <td>HGNC:26927</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>1</td>\n",
       "      <td>Dec 30, 2019</td>\n",
       "      <td>267606829</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>15044</td>\n",
       "      <td>single nucleotide variant</td>\n",
       "      <td>NM_017547.4(FOXRED1):c.694C&gt;T (p.Gln232Ter)</td>\n",
       "      <td>55572</td>\n",
       "      <td>FOXRED1</td>\n",
       "      <td>HGNC:26927</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>1</td>\n",
       "      <td>Dec 30, 2019</td>\n",
       "      <td>267606829</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>15066</td>\n",
       "      <td>single nucleotide variant</td>\n",
       "      <td>NM_001042472.3(ABHD12):c.1054C&gt;T (p.Arg352Ter)</td>\n",
       "      <td>26090</td>\n",
       "      <td>ABHD12</td>\n",
       "      <td>HGNC:15868</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>1</td>\n",
       "      <td>Jan 05, 2022</td>\n",
       "      <td>267606624</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>15066</td>\n",
       "      <td>single nucleotide variant</td>\n",
       "      <td>NM_001042472.3(ABHD12):c.1054C&gt;T (p.Arg352Ter)</td>\n",
       "      <td>26090</td>\n",
       "      <td>ABHD12</td>\n",
       "      <td>HGNC:15868</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>1</td>\n",
       "      <td>Jan 05, 2022</td>\n",
       "      <td>267606624</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52</th>\n",
       "      <td>15068</td>\n",
       "      <td>Microsatellite</td>\n",
       "      <td>NM_138413.4(HOGA1):c.938AGG[2] (p.Glu315del)</td>\n",
       "      <td>112817</td>\n",
       "      <td>HOGA1</td>\n",
       "      <td>HGNC:25155</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>1</td>\n",
       "      <td>Dec 15, 2021</td>\n",
       "      <td>397509360</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3057146</th>\n",
       "      <td>1706616</td>\n",
       "      <td>Deletion</td>\n",
       "      <td>NM_001034853.2(RPGR):c.1225del (p.Arg409fs)</td>\n",
       "      <td>6103</td>\n",
       "      <td>RPGR</td>\n",
       "      <td>HGNC:10295</td>\n",
       "      <td>Pathogenic/Likely pathogenic</td>\n",
       "      <td>1</td>\n",
       "      <td>Apr 14, 2022</td>\n",
       "      <td>-1</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3057261</th>\n",
       "      <td>1706676</td>\n",
       "      <td>Duplication</td>\n",
       "      <td>NM_003322.6(TULP1):c.162dup (p.Thr55fs)</td>\n",
       "      <td>7287</td>\n",
       "      <td>TULP1</td>\n",
       "      <td>HGNC:12423</td>\n",
       "      <td>Pathogenic/Likely pathogenic</td>\n",
       "      <td>1</td>\n",
       "      <td>Dec 30, 2021</td>\n",
       "      <td>-1</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3057262</th>\n",
       "      <td>1706676</td>\n",
       "      <td>Duplication</td>\n",
       "      <td>NM_003322.6(TULP1):c.162dup (p.Thr55fs)</td>\n",
       "      <td>7287</td>\n",
       "      <td>TULP1</td>\n",
       "      <td>HGNC:12423</td>\n",
       "      <td>Pathogenic/Likely pathogenic</td>\n",
       "      <td>1</td>\n",
       "      <td>Dec 30, 2021</td>\n",
       "      <td>-1</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3060703</th>\n",
       "      <td>1708587</td>\n",
       "      <td>Duplication</td>\n",
       "      <td>NM_000152.5(GAA):c.1848dup (p.Val617fs)</td>\n",
       "      <td>2548</td>\n",
       "      <td>GAA</td>\n",
       "      <td>HGNC:4065</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>1</td>\n",
       "      <td>Sep 06, 2022</td>\n",
       "      <td>-1</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3060704</th>\n",
       "      <td>1708587</td>\n",
       "      <td>Duplication</td>\n",
       "      <td>NM_000152.5(GAA):c.1848dup (p.Val617fs)</td>\n",
       "      <td>2548</td>\n",
       "      <td>GAA</td>\n",
       "      <td>HGNC:4065</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>1</td>\n",
       "      <td>Sep 06, 2022</td>\n",
       "      <td>-1</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>78146 rows × 52 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         #AlleleID                       Type  \\\n",
       "6            15044  single nucleotide variant   \n",
       "7            15044  single nucleotide variant   \n",
       "48           15066  single nucleotide variant   \n",
       "49           15066  single nucleotide variant   \n",
       "52           15068             Microsatellite   \n",
       "...            ...                        ...   \n",
       "3057146    1706616                   Deletion   \n",
       "3057261    1706676                Duplication   \n",
       "3057262    1706676                Duplication   \n",
       "3060703    1708587                Duplication   \n",
       "3060704    1708587                Duplication   \n",
       "\n",
       "                                                   Name  GeneID GeneSymbol  \\\n",
       "6           NM_017547.4(FOXRED1):c.694C>T (p.Gln232Ter)   55572    FOXRED1   \n",
       "7           NM_017547.4(FOXRED1):c.694C>T (p.Gln232Ter)   55572    FOXRED1   \n",
       "48       NM_001042472.3(ABHD12):c.1054C>T (p.Arg352Ter)   26090     ABHD12   \n",
       "49       NM_001042472.3(ABHD12):c.1054C>T (p.Arg352Ter)   26090     ABHD12   \n",
       "52         NM_138413.4(HOGA1):c.938AGG[2] (p.Glu315del)  112817      HOGA1   \n",
       "...                                                 ...     ...        ...   \n",
       "3057146     NM_001034853.2(RPGR):c.1225del (p.Arg409fs)    6103       RPGR   \n",
       "3057261         NM_003322.6(TULP1):c.162dup (p.Thr55fs)    7287      TULP1   \n",
       "3057262         NM_003322.6(TULP1):c.162dup (p.Thr55fs)    7287      TULP1   \n",
       "3060703         NM_000152.5(GAA):c.1848dup (p.Val617fs)    2548        GAA   \n",
       "3060704         NM_000152.5(GAA):c.1848dup (p.Val617fs)    2548        GAA   \n",
       "\n",
       "            HGNC_ID          ClinicalSignificance  ClinSigSimple  \\\n",
       "6        HGNC:26927                    Pathogenic              1   \n",
       "7        HGNC:26927                    Pathogenic              1   \n",
       "48       HGNC:15868                    Pathogenic              1   \n",
       "49       HGNC:15868                    Pathogenic              1   \n",
       "52       HGNC:25155                    Pathogenic              1   \n",
       "...             ...                           ...            ...   \n",
       "3057146  HGNC:10295  Pathogenic/Likely pathogenic              1   \n",
       "3057261  HGNC:12423  Pathogenic/Likely pathogenic              1   \n",
       "3057262  HGNC:12423  Pathogenic/Likely pathogenic              1   \n",
       "3060703   HGNC:4065                    Pathogenic              1   \n",
       "3060704   HGNC:4065                    Pathogenic              1   \n",
       "\n",
       "        LastEvaluated  RS# (dbSNP)  ... phenotype_DCM phenotype_HCM  \\\n",
       "6        Dec 30, 2019    267606829  ...             0             0   \n",
       "7        Dec 30, 2019    267606829  ...             0             0   \n",
       "48       Jan 05, 2022    267606624  ...             0             0   \n",
       "49       Jan 05, 2022    267606624  ...             0             0   \n",
       "52       Dec 15, 2021    397509360  ...             0             0   \n",
       "...               ...          ...  ...           ...           ...   \n",
       "3057146  Apr 14, 2022           -1  ...             0             0   \n",
       "3057261  Dec 30, 2021           -1  ...             0             0   \n",
       "3057262  Dec 30, 2021           -1  ...             0             0   \n",
       "3060703  Sep 06, 2022           -1  ...             0             0   \n",
       "3060704  Sep 06, 2022           -1  ...             0             0   \n",
       "\n",
       "        phenotype_LVNC phenotype_Venous anomaly phenotype_Pulmonary artery  \\\n",
       "6                    0                        0                          0   \n",
       "7                    0                        0                          0   \n",
       "48                   0                        0                          0   \n",
       "49                   0                        0                          0   \n",
       "52                   0                        0                          0   \n",
       "...                ...                      ...                        ...   \n",
       "3057146              0                        0                          0   \n",
       "3057261              0                        0                          0   \n",
       "3057262              0                        0                          0   \n",
       "3060703              0                        0                          0   \n",
       "3060704              0                        0                          0   \n",
       "\n",
       "        phenotype_Aortic arch phenotype_Aorta phenotype_Coronary  \\\n",
       "6                           0               0                  0   \n",
       "7                           0               0                  0   \n",
       "48                          0               0                  0   \n",
       "49                          0               0                  0   \n",
       "52                          0               0                  0   \n",
       "...                       ...             ...                ...   \n",
       "3057146                     0               0                  0   \n",
       "3057261                     0               0                  0   \n",
       "3057262                     0               0                  0   \n",
       "3060703                     0               0                  0   \n",
       "3060704                     0               0                  0   \n",
       "\n",
       "        phenotype_VA valves  phenotype_AV valves  \n",
       "6                         0                    0  \n",
       "7                         0                    0  \n",
       "48                        0                    0  \n",
       "49                        0                    0  \n",
       "52                        0                    0  \n",
       "...                     ...                  ...  \n",
       "3057146                   0                    0  \n",
       "3057261                   0                    0  \n",
       "3057262                   0                    0  \n",
       "3060703                   0                    0  \n",
       "3060704                   0                    0  \n",
       "\n",
       "[78146 rows x 52 columns]"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Make new columns indicating whether a phenotype is present\n",
    "phenotype_column='PhenotypeList'\n",
    "\n",
    "for phenotype in range(len(phenotype_dict)):\n",
    "    clinvar_summary_df['phenotype_'+str(list(phenotype_dict.keys())[phenotype])]=0\n",
    "    clinvar_summary_df['phenotype_'+str(list(phenotype_dict.keys())[phenotype])][clinvar_summary_df[phenotype_column].str.contains(case=False,pat='|'.join(list(phenotype_dict.values())[phenotype]))]=1\n",
    "\n",
    "clinvar_summary_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "changed-minute",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_5443/2597537464.py:3: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  clinvar_summary_df['TF'][clinvar_summary_df.index.isin(tf_list)]=1 # add a column indicating whether TF or not\n"
     ]
    }
   ],
   "source": [
    "# Add a TF column\n",
    "clinvar_summary_df['TF']=0\n",
    "clinvar_summary_df['TF'][clinvar_summary_df.index.isin(tf_list)]=1 # add a column indicating whether TF or not"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "related-canvas",
   "metadata": {},
   "source": [
    "# Review"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "quantitative-python",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "Heterotaxy\n",
      "ZIC3          4\n",
      "MMP21         4\n",
      "CERS1;GDF1    2\n",
      "NODAL         2\n",
      "Name: GeneSymbol, dtype: int64\n",
      "\n",
      "\n",
      "ASD\n",
      "NKX2-5    6\n",
      "ACTC1     4\n",
      "TAB2      2\n",
      "Name: GeneSymbol, dtype: int64\n",
      "\n",
      "\n",
      "VSD\n",
      "Series([], Name: GeneSymbol, dtype: int64)\n",
      "\n",
      "\n",
      "AVSD\n",
      "Series([], Name: GeneSymbol, dtype: int64)\n",
      "\n",
      "\n",
      "Malformation of outflow tracts\n",
      "MED13L    6\n",
      "JAG1      4\n",
      "FOXC1     2\n",
      "ROBO1     2\n",
      "Name: GeneSymbol, dtype: int64\n",
      "\n",
      "\n",
      "Functional single ventricle\n",
      "Series([], Name: GeneSymbol, dtype: int64)\n",
      "\n",
      "\n",
      "Cardiomyopathy\n",
      "TTN       664\n",
      "MYBPC3    526\n",
      "MYH7      234\n",
      "DSP       176\n",
      "LMNA      118\n",
      "         ... \n",
      "MT-TI       2\n",
      "COQ9        2\n",
      "MT-TL1      2\n",
      "MT-TK       2\n",
      "PRDM16      1\n",
      "Name: GeneSymbol, Length: 61, dtype: int64\n",
      "\n",
      "\n",
      "DCM\n",
      "TTN                 656\n",
      "LMNA                 96\n",
      "FLNC                 52\n",
      "MYH7                 46\n",
      "TNNT2                42\n",
      "BAG3                 36\n",
      "DSP                  30\n",
      "DMD                  24\n",
      "SCN5A                14\n",
      "RBM20                14\n",
      "DES                  12\n",
      "ABCC9                10\n",
      "TPM1                 10\n",
      "MYBPC3               10\n",
      "ACTC1                 8\n",
      "RAF1                  8\n",
      "TNNI3                 6\n",
      "FKTN                  4\n",
      "PSEN1                 4\n",
      "SGCD                  4\n",
      "PLN                   4\n",
      "TMEM43                2\n",
      "NEXN                  2\n",
      "MYPN                  2\n",
      "NKX2-5                2\n",
      "TCAP                  2\n",
      "NCAPH2;SCO2;TYMP      2\n",
      "EYA4                  2\n",
      "ACTN2                 2\n",
      "DSG2                  2\n",
      "SDHA                  2\n",
      "SGCB                  2\n",
      "MT-TI                 2\n",
      "CRYAB                 2\n",
      "TNNC1                 2\n",
      "PPP1R13L              2\n",
      "Name: GeneSymbol, dtype: int64\n",
      "\n",
      "\n",
      "HCM\n",
      "MYBPC3     522\n",
      "MYH7       228\n",
      "FLNC        52\n",
      "TNNT2       46\n",
      "TNNI3       34\n",
      "MYL2        12\n",
      "TCAP        12\n",
      "MTO1        12\n",
      "TPM1        12\n",
      "PRKAG2      10\n",
      "LAMP2        8\n",
      "GLA          8\n",
      "RAF1         8\n",
      "MYL3         6\n",
      "ALPK3        6\n",
      "ACTC1        6\n",
      "TTN          4\n",
      "PTPN11       4\n",
      "LMNA         4\n",
      "ACTN2        2\n",
      "NEXN         2\n",
      "CASR         2\n",
      "CACNA1C      2\n",
      "COQ9         2\n",
      "DES          2\n",
      "MRPL44       2\n",
      "TMEM43       2\n",
      "PLN          2\n",
      "TNNC1        2\n",
      "MT-TI        2\n",
      "MYO6         2\n",
      "TRIM63       2\n",
      "Name: GeneSymbol, dtype: int64\n",
      "\n",
      "\n",
      "LVNC\n",
      "MYBPC3    42\n",
      "MYH7      18\n",
      "TTN        8\n",
      "ACTC1      2\n",
      "TPM1       2\n",
      "LMNA       2\n",
      "RBM20      2\n",
      "DSP        2\n",
      "PRDM16     1\n",
      "Name: GeneSymbol, dtype: int64\n",
      "\n",
      "\n",
      "Venous anomaly\n",
      "Series([], Name: GeneSymbol, dtype: int64)\n",
      "\n",
      "\n",
      "Pulmonary artery\n",
      "Series([], Name: GeneSymbol, dtype: int64)\n",
      "\n",
      "\n",
      "Aortic arch\n",
      "FLNA    2\n",
      "Name: GeneSymbol, dtype: int64\n",
      "\n",
      "\n",
      "Aorta\n",
      "FBN1       736\n",
      "SMAD3       40\n",
      "COL3A1      22\n",
      "ACTA2       22\n",
      "TGFBR2      20\n",
      "TGFBR1      16\n",
      "MYH11       10\n",
      "TGFB2        6\n",
      "LOX          4\n",
      "PRKG1        2\n",
      "SLC2A10      2\n",
      "MFAP5        2\n",
      "TGFB3        2\n",
      "NOTCH1       2\n",
      "Name: GeneSymbol, dtype: int64\n",
      "\n",
      "\n",
      "Coronary\n",
      "Series([], Name: GeneSymbol, dtype: int64)\n",
      "\n",
      "\n",
      "VA valves\n",
      "NF1     48\n",
      "ELN     26\n",
      "SOS1     2\n",
      "TAB2     2\n",
      "Name: GeneSymbol, dtype: int64\n",
      "\n",
      "\n",
      "AV valves\n",
      "Series([], Name: GeneSymbol, dtype: int64)\n"
     ]
    }
   ],
   "source": [
    "for phenotype in range(len(phenotype_dict)):\n",
    "    print('\\n')\n",
    "    print(str(list(phenotype_dict.keys())[phenotype]))\n",
    "    print(clinvar_summary_df['GeneSymbol'][clinvar_summary_df['phenotype_'+str(list(phenotype_dict.keys())[phenotype])]==1].value_counts())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "greater-humor",
   "metadata": {},
   "source": [
    "# Tidy for export"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "dried-warehouse",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#AlleleID</th>\n",
       "      <th>Type</th>\n",
       "      <th>Name</th>\n",
       "      <th>GeneID</th>\n",
       "      <th>HGNC_ID</th>\n",
       "      <th>ClinicalSignificance</th>\n",
       "      <th>ClinSigSimple</th>\n",
       "      <th>LastEvaluated</th>\n",
       "      <th>RS# (dbSNP)</th>\n",
       "      <th>nsv/esv (dbVar)</th>\n",
       "      <th>...</th>\n",
       "      <th>phenotype_HCM</th>\n",
       "      <th>phenotype_LVNC</th>\n",
       "      <th>phenotype_Venous anomaly</th>\n",
       "      <th>phenotype_Pulmonary artery</th>\n",
       "      <th>phenotype_Aortic arch</th>\n",
       "      <th>phenotype_Aorta</th>\n",
       "      <th>phenotype_Coronary</th>\n",
       "      <th>phenotype_VA valves</th>\n",
       "      <th>phenotype_AV valves</th>\n",
       "      <th>TF</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GeneSymbol</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>FOXRED1</th>\n",
       "      <td>15044</td>\n",
       "      <td>single nucleotide variant</td>\n",
       "      <td>NM_017547.4(FOXRED1):c.694C&gt;T (p.Gln232Ter)</td>\n",
       "      <td>55572</td>\n",
       "      <td>HGNC:26927</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>1</td>\n",
       "      <td>Dec 30, 2019</td>\n",
       "      <td>267606829</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FOXRED1</th>\n",
       "      <td>15044</td>\n",
       "      <td>single nucleotide variant</td>\n",
       "      <td>NM_017547.4(FOXRED1):c.694C&gt;T (p.Gln232Ter)</td>\n",
       "      <td>55572</td>\n",
       "      <td>HGNC:26927</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>1</td>\n",
       "      <td>Dec 30, 2019</td>\n",
       "      <td>267606829</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ABHD12</th>\n",
       "      <td>15066</td>\n",
       "      <td>single nucleotide variant</td>\n",
       "      <td>NM_001042472.3(ABHD12):c.1054C&gt;T (p.Arg352Ter)</td>\n",
       "      <td>26090</td>\n",
       "      <td>HGNC:15868</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>1</td>\n",
       "      <td>Jan 05, 2022</td>\n",
       "      <td>267606624</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ABHD12</th>\n",
       "      <td>15066</td>\n",
       "      <td>single nucleotide variant</td>\n",
       "      <td>NM_001042472.3(ABHD12):c.1054C&gt;T (p.Arg352Ter)</td>\n",
       "      <td>26090</td>\n",
       "      <td>HGNC:15868</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>1</td>\n",
       "      <td>Jan 05, 2022</td>\n",
       "      <td>267606624</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HOGA1</th>\n",
       "      <td>15068</td>\n",
       "      <td>Microsatellite</td>\n",
       "      <td>NM_138413.4(HOGA1):c.938AGG[2] (p.Glu315del)</td>\n",
       "      <td>112817</td>\n",
       "      <td>HGNC:25155</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>1</td>\n",
       "      <td>Dec 15, 2021</td>\n",
       "      <td>397509360</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RPGR</th>\n",
       "      <td>1706616</td>\n",
       "      <td>Deletion</td>\n",
       "      <td>NM_001034853.2(RPGR):c.1225del (p.Arg409fs)</td>\n",
       "      <td>6103</td>\n",
       "      <td>HGNC:10295</td>\n",
       "      <td>Pathogenic/Likely pathogenic</td>\n",
       "      <td>1</td>\n",
       "      <td>Apr 14, 2022</td>\n",
       "      <td>-1</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TULP1</th>\n",
       "      <td>1706676</td>\n",
       "      <td>Duplication</td>\n",
       "      <td>NM_003322.6(TULP1):c.162dup (p.Thr55fs)</td>\n",
       "      <td>7287</td>\n",
       "      <td>HGNC:12423</td>\n",
       "      <td>Pathogenic/Likely pathogenic</td>\n",
       "      <td>1</td>\n",
       "      <td>Dec 30, 2021</td>\n",
       "      <td>-1</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TULP1</th>\n",
       "      <td>1706676</td>\n",
       "      <td>Duplication</td>\n",
       "      <td>NM_003322.6(TULP1):c.162dup (p.Thr55fs)</td>\n",
       "      <td>7287</td>\n",
       "      <td>HGNC:12423</td>\n",
       "      <td>Pathogenic/Likely pathogenic</td>\n",
       "      <td>1</td>\n",
       "      <td>Dec 30, 2021</td>\n",
       "      <td>-1</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GAA</th>\n",
       "      <td>1708587</td>\n",
       "      <td>Duplication</td>\n",
       "      <td>NM_000152.5(GAA):c.1848dup (p.Val617fs)</td>\n",
       "      <td>2548</td>\n",
       "      <td>HGNC:4065</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>1</td>\n",
       "      <td>Sep 06, 2022</td>\n",
       "      <td>-1</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GAA</th>\n",
       "      <td>1708587</td>\n",
       "      <td>Duplication</td>\n",
       "      <td>NM_000152.5(GAA):c.1848dup (p.Val617fs)</td>\n",
       "      <td>2548</td>\n",
       "      <td>HGNC:4065</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>1</td>\n",
       "      <td>Sep 06, 2022</td>\n",
       "      <td>-1</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>78146 rows × 52 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            #AlleleID                       Type  \\\n",
       "GeneSymbol                                         \n",
       "FOXRED1         15044  single nucleotide variant   \n",
       "FOXRED1         15044  single nucleotide variant   \n",
       "ABHD12          15066  single nucleotide variant   \n",
       "ABHD12          15066  single nucleotide variant   \n",
       "HOGA1           15068             Microsatellite   \n",
       "...               ...                        ...   \n",
       "RPGR          1706616                   Deletion   \n",
       "TULP1         1706676                Duplication   \n",
       "TULP1         1706676                Duplication   \n",
       "GAA           1708587                Duplication   \n",
       "GAA           1708587                Duplication   \n",
       "\n",
       "                                                      Name  GeneID  \\\n",
       "GeneSymbol                                                           \n",
       "FOXRED1        NM_017547.4(FOXRED1):c.694C>T (p.Gln232Ter)   55572   \n",
       "FOXRED1        NM_017547.4(FOXRED1):c.694C>T (p.Gln232Ter)   55572   \n",
       "ABHD12      NM_001042472.3(ABHD12):c.1054C>T (p.Arg352Ter)   26090   \n",
       "ABHD12      NM_001042472.3(ABHD12):c.1054C>T (p.Arg352Ter)   26090   \n",
       "HOGA1         NM_138413.4(HOGA1):c.938AGG[2] (p.Glu315del)  112817   \n",
       "...                                                    ...     ...   \n",
       "RPGR           NM_001034853.2(RPGR):c.1225del (p.Arg409fs)    6103   \n",
       "TULP1              NM_003322.6(TULP1):c.162dup (p.Thr55fs)    7287   \n",
       "TULP1              NM_003322.6(TULP1):c.162dup (p.Thr55fs)    7287   \n",
       "GAA                NM_000152.5(GAA):c.1848dup (p.Val617fs)    2548   \n",
       "GAA                NM_000152.5(GAA):c.1848dup (p.Val617fs)    2548   \n",
       "\n",
       "               HGNC_ID          ClinicalSignificance  ClinSigSimple  \\\n",
       "GeneSymbol                                                            \n",
       "FOXRED1     HGNC:26927                    Pathogenic              1   \n",
       "FOXRED1     HGNC:26927                    Pathogenic              1   \n",
       "ABHD12      HGNC:15868                    Pathogenic              1   \n",
       "ABHD12      HGNC:15868                    Pathogenic              1   \n",
       "HOGA1       HGNC:25155                    Pathogenic              1   \n",
       "...                ...                           ...            ...   \n",
       "RPGR        HGNC:10295  Pathogenic/Likely pathogenic              1   \n",
       "TULP1       HGNC:12423  Pathogenic/Likely pathogenic              1   \n",
       "TULP1       HGNC:12423  Pathogenic/Likely pathogenic              1   \n",
       "GAA          HGNC:4065                    Pathogenic              1   \n",
       "GAA          HGNC:4065                    Pathogenic              1   \n",
       "\n",
       "           LastEvaluated  RS# (dbSNP) nsv/esv (dbVar)  ... phenotype_HCM  \\\n",
       "GeneSymbol                                             ...                 \n",
       "FOXRED1     Dec 30, 2019    267606829               -  ...             0   \n",
       "FOXRED1     Dec 30, 2019    267606829               -  ...             0   \n",
       "ABHD12      Jan 05, 2022    267606624               -  ...             0   \n",
       "ABHD12      Jan 05, 2022    267606624               -  ...             0   \n",
       "HOGA1       Dec 15, 2021    397509360               -  ...             0   \n",
       "...                  ...          ...             ...  ...           ...   \n",
       "RPGR        Apr 14, 2022           -1               -  ...             0   \n",
       "TULP1       Dec 30, 2021           -1               -  ...             0   \n",
       "TULP1       Dec 30, 2021           -1               -  ...             0   \n",
       "GAA         Sep 06, 2022           -1               -  ...             0   \n",
       "GAA         Sep 06, 2022           -1               -  ...             0   \n",
       "\n",
       "           phenotype_LVNC phenotype_Venous anomaly phenotype_Pulmonary artery  \\\n",
       "GeneSymbol                                                                      \n",
       "FOXRED1                 0                        0                          0   \n",
       "FOXRED1                 0                        0                          0   \n",
       "ABHD12                  0                        0                          0   \n",
       "ABHD12                  0                        0                          0   \n",
       "HOGA1                   0                        0                          0   \n",
       "...                   ...                      ...                        ...   \n",
       "RPGR                    0                        0                          0   \n",
       "TULP1                   0                        0                          0   \n",
       "TULP1                   0                        0                          0   \n",
       "GAA                     0                        0                          0   \n",
       "GAA                     0                        0                          0   \n",
       "\n",
       "           phenotype_Aortic arch phenotype_Aorta phenotype_Coronary  \\\n",
       "GeneSymbol                                                            \n",
       "FOXRED1                        0               0                  0   \n",
       "FOXRED1                        0               0                  0   \n",
       "ABHD12                         0               0                  0   \n",
       "ABHD12                         0               0                  0   \n",
       "HOGA1                          0               0                  0   \n",
       "...                          ...             ...                ...   \n",
       "RPGR                           0               0                  0   \n",
       "TULP1                          0               0                  0   \n",
       "TULP1                          0               0                  0   \n",
       "GAA                            0               0                  0   \n",
       "GAA                            0               0                  0   \n",
       "\n",
       "           phenotype_VA valves  phenotype_AV valves  TF  \n",
       "GeneSymbol                                               \n",
       "FOXRED1                      0                    0   0  \n",
       "FOXRED1                      0                    0   0  \n",
       "ABHD12                       0                    0   0  \n",
       "ABHD12                       0                    0   0  \n",
       "HOGA1                        0                    0   0  \n",
       "...                        ...                  ...  ..  \n",
       "RPGR                         0                    0   0  \n",
       "TULP1                        0                    0   0  \n",
       "TULP1                        0                    0   0  \n",
       "GAA                          0                    0   0  \n",
       "GAA                          0                    0   0  \n",
       "\n",
       "[78146 rows x 52 columns]"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clinvar_summary_df.set_index('GeneSymbol',inplace=True)\n",
    "clinvar_summary_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "novel-hindu",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(3349, 52)\n",
      "(3349, 52)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#AlleleID</th>\n",
       "      <th>Type</th>\n",
       "      <th>Name</th>\n",
       "      <th>GeneID</th>\n",
       "      <th>HGNC_ID</th>\n",
       "      <th>ClinicalSignificance</th>\n",
       "      <th>ClinSigSimple</th>\n",
       "      <th>LastEvaluated</th>\n",
       "      <th>RS# (dbSNP)</th>\n",
       "      <th>nsv/esv (dbVar)</th>\n",
       "      <th>...</th>\n",
       "      <th>phenotype_LVNC</th>\n",
       "      <th>phenotype_Venous anomaly</th>\n",
       "      <th>phenotype_Pulmonary artery</th>\n",
       "      <th>phenotype_Aortic arch</th>\n",
       "      <th>phenotype_Aorta</th>\n",
       "      <th>phenotype_Coronary</th>\n",
       "      <th>phenotype_VA valves</th>\n",
       "      <th>phenotype_AV valves</th>\n",
       "      <th>TF</th>\n",
       "      <th>source</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GeneSymbol</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>RBM20</th>\n",
       "      <td>15307</td>\n",
       "      <td>single nucleotide variant</td>\n",
       "      <td>NM_001134363.3(RBM20):c.1913C&gt;T (p.Pro638Leu)</td>\n",
       "      <td>282996</td>\n",
       "      <td>HGNC:27424</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>1</td>\n",
       "      <td>Jul 26, 2022</td>\n",
       "      <td>267607003</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>ClinVar</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RBM20</th>\n",
       "      <td>15307</td>\n",
       "      <td>single nucleotide variant</td>\n",
       "      <td>NM_001134363.3(RBM20):c.1913C&gt;T (p.Pro638Leu)</td>\n",
       "      <td>282996</td>\n",
       "      <td>HGNC:27424</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>1</td>\n",
       "      <td>Jul 26, 2022</td>\n",
       "      <td>267607003</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>ClinVar</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RBM20</th>\n",
       "      <td>15308</td>\n",
       "      <td>single nucleotide variant</td>\n",
       "      <td>NM_001134363.3(RBM20):c.1901G&gt;A (p.Arg634Gln)</td>\n",
       "      <td>282996</td>\n",
       "      <td>HGNC:27424</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>1</td>\n",
       "      <td>Mar 18, 2022</td>\n",
       "      <td>267607001</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>ClinVar</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RBM20</th>\n",
       "      <td>15308</td>\n",
       "      <td>single nucleotide variant</td>\n",
       "      <td>NM_001134363.3(RBM20):c.1901G&gt;A (p.Arg634Gln)</td>\n",
       "      <td>282996</td>\n",
       "      <td>HGNC:27424</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>1</td>\n",
       "      <td>Mar 18, 2022</td>\n",
       "      <td>267607001</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>ClinVar</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RBM20</th>\n",
       "      <td>15309</td>\n",
       "      <td>single nucleotide variant</td>\n",
       "      <td>NM_001134363.3(RBM20):c.1906C&gt;A (p.Arg636Ser)</td>\n",
       "      <td>282996</td>\n",
       "      <td>HGNC:27424</td>\n",
       "      <td>Pathogenic/Likely pathogenic</td>\n",
       "      <td>1</td>\n",
       "      <td>Jan 06, 2022</td>\n",
       "      <td>267607002</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>ClinVar</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTN</th>\n",
       "      <td>1464444</td>\n",
       "      <td>Duplication</td>\n",
       "      <td>NM_001267550.2(TTN):c.68286_68289dup (p.Trp227...</td>\n",
       "      <td>7273</td>\n",
       "      <td>HGNC:12403</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>1</td>\n",
       "      <td>Apr 14, 2022</td>\n",
       "      <td>-1</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>ClinVar</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTN</th>\n",
       "      <td>1464767</td>\n",
       "      <td>single nucleotide variant</td>\n",
       "      <td>NM_001267550.2(TTN):c.66628C&gt;T (p.Gln22210Ter)</td>\n",
       "      <td>7273</td>\n",
       "      <td>HGNC:12403</td>\n",
       "      <td>Pathogenic/Likely pathogenic</td>\n",
       "      <td>1</td>\n",
       "      <td>Mar 07, 2022</td>\n",
       "      <td>-1</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>ClinVar</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTN</th>\n",
       "      <td>1464767</td>\n",
       "      <td>single nucleotide variant</td>\n",
       "      <td>NM_001267550.2(TTN):c.66628C&gt;T (p.Gln22210Ter)</td>\n",
       "      <td>7273</td>\n",
       "      <td>HGNC:12403</td>\n",
       "      <td>Pathogenic/Likely pathogenic</td>\n",
       "      <td>1</td>\n",
       "      <td>Mar 07, 2022</td>\n",
       "      <td>-1</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>ClinVar</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PKP2</th>\n",
       "      <td>1487925</td>\n",
       "      <td>Deletion</td>\n",
       "      <td>NM_001005242.3(PKP2):c.1264_1265del (p.Leu422fs)</td>\n",
       "      <td>5318</td>\n",
       "      <td>HGNC:9024</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>1</td>\n",
       "      <td>Jun 01, 2022</td>\n",
       "      <td>-1</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>ClinVar</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PKP2</th>\n",
       "      <td>1487925</td>\n",
       "      <td>Deletion</td>\n",
       "      <td>NM_001005242.3(PKP2):c.1264_1265del (p.Leu422fs)</td>\n",
       "      <td>5318</td>\n",
       "      <td>HGNC:9024</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>1</td>\n",
       "      <td>Jun 01, 2022</td>\n",
       "      <td>-1</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>ClinVar</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3349 rows × 53 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            #AlleleID                       Type  \\\n",
       "GeneSymbol                                         \n",
       "RBM20           15307  single nucleotide variant   \n",
       "RBM20           15307  single nucleotide variant   \n",
       "RBM20           15308  single nucleotide variant   \n",
       "RBM20           15308  single nucleotide variant   \n",
       "RBM20           15309  single nucleotide variant   \n",
       "...               ...                        ...   \n",
       "TTN           1464444                Duplication   \n",
       "TTN           1464767  single nucleotide variant   \n",
       "TTN           1464767  single nucleotide variant   \n",
       "PKP2          1487925                   Deletion   \n",
       "PKP2          1487925                   Deletion   \n",
       "\n",
       "                                                         Name  GeneID  \\\n",
       "GeneSymbol                                                              \n",
       "RBM20           NM_001134363.3(RBM20):c.1913C>T (p.Pro638Leu)  282996   \n",
       "RBM20           NM_001134363.3(RBM20):c.1913C>T (p.Pro638Leu)  282996   \n",
       "RBM20           NM_001134363.3(RBM20):c.1901G>A (p.Arg634Gln)  282996   \n",
       "RBM20           NM_001134363.3(RBM20):c.1901G>A (p.Arg634Gln)  282996   \n",
       "RBM20           NM_001134363.3(RBM20):c.1906C>A (p.Arg636Ser)  282996   \n",
       "...                                                       ...     ...   \n",
       "TTN         NM_001267550.2(TTN):c.68286_68289dup (p.Trp227...    7273   \n",
       "TTN            NM_001267550.2(TTN):c.66628C>T (p.Gln22210Ter)    7273   \n",
       "TTN            NM_001267550.2(TTN):c.66628C>T (p.Gln22210Ter)    7273   \n",
       "PKP2         NM_001005242.3(PKP2):c.1264_1265del (p.Leu422fs)    5318   \n",
       "PKP2         NM_001005242.3(PKP2):c.1264_1265del (p.Leu422fs)    5318   \n",
       "\n",
       "               HGNC_ID          ClinicalSignificance  ClinSigSimple  \\\n",
       "GeneSymbol                                                            \n",
       "RBM20       HGNC:27424                    Pathogenic              1   \n",
       "RBM20       HGNC:27424                    Pathogenic              1   \n",
       "RBM20       HGNC:27424                    Pathogenic              1   \n",
       "RBM20       HGNC:27424                    Pathogenic              1   \n",
       "RBM20       HGNC:27424  Pathogenic/Likely pathogenic              1   \n",
       "...                ...                           ...            ...   \n",
       "TTN         HGNC:12403                    Pathogenic              1   \n",
       "TTN         HGNC:12403  Pathogenic/Likely pathogenic              1   \n",
       "TTN         HGNC:12403  Pathogenic/Likely pathogenic              1   \n",
       "PKP2         HGNC:9024                    Pathogenic              1   \n",
       "PKP2         HGNC:9024                    Pathogenic              1   \n",
       "\n",
       "           LastEvaluated  RS# (dbSNP) nsv/esv (dbVar)  ... phenotype_LVNC  \\\n",
       "GeneSymbol                                             ...                  \n",
       "RBM20       Jul 26, 2022    267607003               -  ...              0   \n",
       "RBM20       Jul 26, 2022    267607003               -  ...              0   \n",
       "RBM20       Mar 18, 2022    267607001               -  ...              0   \n",
       "RBM20       Mar 18, 2022    267607001               -  ...              0   \n",
       "RBM20       Jan 06, 2022    267607002               -  ...              0   \n",
       "...                  ...          ...             ...  ...            ...   \n",
       "TTN         Apr 14, 2022           -1               -  ...              0   \n",
       "TTN         Mar 07, 2022           -1               -  ...              0   \n",
       "TTN         Mar 07, 2022           -1               -  ...              0   \n",
       "PKP2        Jun 01, 2022           -1               -  ...              0   \n",
       "PKP2        Jun 01, 2022           -1               -  ...              0   \n",
       "\n",
       "           phenotype_Venous anomaly phenotype_Pulmonary artery  \\\n",
       "GeneSymbol                                                       \n",
       "RBM20                             0                          0   \n",
       "RBM20                             0                          0   \n",
       "RBM20                             0                          0   \n",
       "RBM20                             0                          0   \n",
       "RBM20                             0                          0   \n",
       "...                             ...                        ...   \n",
       "TTN                               0                          0   \n",
       "TTN                               0                          0   \n",
       "TTN                               0                          0   \n",
       "PKP2                              0                          0   \n",
       "PKP2                              0                          0   \n",
       "\n",
       "           phenotype_Aortic arch phenotype_Aorta phenotype_Coronary  \\\n",
       "GeneSymbol                                                            \n",
       "RBM20                          0               0                  0   \n",
       "RBM20                          0               0                  0   \n",
       "RBM20                          0               0                  0   \n",
       "RBM20                          0               0                  0   \n",
       "RBM20                          0               0                  0   \n",
       "...                          ...             ...                ...   \n",
       "TTN                            0               0                  0   \n",
       "TTN                            0               0                  0   \n",
       "TTN                            0               0                  0   \n",
       "PKP2                           0               0                  0   \n",
       "PKP2                           0               0                  0   \n",
       "\n",
       "           phenotype_VA valves phenotype_AV valves  TF   source  \n",
       "GeneSymbol                                                       \n",
       "RBM20                        0                   0   0  ClinVar  \n",
       "RBM20                        0                   0   0  ClinVar  \n",
       "RBM20                        0                   0   0  ClinVar  \n",
       "RBM20                        0                   0   0  ClinVar  \n",
       "RBM20                        0                   0   0  ClinVar  \n",
       "...                        ...                 ...  ..      ...  \n",
       "TTN                          0                   0   0  ClinVar  \n",
       "TTN                          0                   0   0  ClinVar  \n",
       "TTN                          0                   0   0  ClinVar  \n",
       "PKP2                         0                   0   0  ClinVar  \n",
       "PKP2                         0                   0   0  ClinVar  \n",
       "\n",
       "[3349 rows x 53 columns]"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Remove rows where there is no association with a cardiac phenotype\n",
    "cols=clinvar_summary_df.columns.tolist()\n",
    "phen_cols=[x for x in cols if x.startswith('phenotype_')]\n",
    "print(clinvar_summary_df.shape)\n",
    "clinvar_summary_df=clinvar_summary_df.loc[~(clinvar_summary_df[phen_cols]==0).all(axis=1)]\n",
    "print(clinvar_summary_df.shape)\n",
    "clinvar_summary_df['source']='ClinVar'\n",
    "clinvar_summary_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "brutal-terrain",
   "metadata": {},
   "outputs": [],
   "source": [
    "clinvar_summary_df.to_csv('/home/jovyan/data/ClinVar_with_phenotype_annotation.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "diverse-engineering",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "bioconda",
   "language": "python",
   "name": "bioconda"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
