{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "noticed-princess",
   "metadata": {},
   "source": [
    "# Prepare Compass inputs, random groups within cell states"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "tough-villa",
   "metadata": {},
   "source": [
    "Due to large numbers of micropool clusters being contaminated with cells from other annotated groups, we will divide up cells within each celltype into groups of 20 (CM) and 50 (all cells)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "competitive-relations",
   "metadata": {},
   "source": [
    "## Setup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "union-quebec",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/jovyan/my-conda-envs/scanpy/lib/python3.9/site-packages/tqdm/auto.py:22: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html\n",
      "  from .autonotebook import tqdm as notebook_tqdm\n"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import scanpy as sc\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "from matplotlib.pyplot import rc_context\n",
    "import bbknn\n",
    "import re\n",
    "import sctk"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "persistent-travel",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "scanpy==1.9.1 anndata==0.8.0 umap==0.5.3 numpy==1.22.4 scipy==1.8.1 pandas==1.4.3 scikit-learn==1.1.1 statsmodels==0.13.2 python-igraph==0.9.11 pynndescent==0.5.7\n"
     ]
    }
   ],
   "source": [
    "sc.settings.verbosity = 0            # verbosity: errors (0), warnings (1), info (2), hints (3)\n",
    "sc.logging.print_header()\n",
    "sc.settings.set_figure_params(dpi=80, facecolor='white', color_map = 'Reds', figsize = (5, 5),dpi_save=300)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "advisory-synthesis",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Perform pseudobulk on sc data\n",
    "\n",
    "import scipy.sparse as sp\n",
    "\n",
    "def pseudo_bulk(\n",
    "        adata, groupby, use_rep='X', highly_variable=False, FUN=np.mean):\n",
    "    \"\"\"Make pseudo bulk data from grouped sc data\n",
    "    \"\"\"\n",
    "    if adata.obs[groupby].dtype.name == 'category':\n",
    "        group_attr = adata.obs[groupby].values\n",
    "        groups = adata.obs[groupby].cat.categories.values\n",
    "    else:\n",
    "        group_attr = adata.obs[groupby].astype(str).values\n",
    "        groups = np.unique(group_attr)\n",
    "    n_level = len(groups)\n",
    "    if highly_variable:\n",
    "        if isinstance(highly_variable, (list, tuple)):\n",
    "            if use_rep == 'raw':\n",
    "                k_hv = adata.raw.var_names.isin(highly_variable)\n",
    "            else:\n",
    "                k_hv = adata.var_names.isin(highly_variable)\n",
    "        else:\n",
    "            k_hv = adata.var['highly_variable'].values\n",
    "    if use_rep == 'X':\n",
    "        x = adata.X\n",
    "        features = adata.var_names.values\n",
    "        if highly_variable:\n",
    "            x = x[:, k_hv]\n",
    "            features = features[k_hv]\n",
    "    elif use_rep == 'raw':\n",
    "        x = adata.raw.X\n",
    "        features = adata.raw.var_names.values\n",
    "        if highly_variable:\n",
    "            x = x[:, k_hv]\n",
    "            features = features[k_hv]\n",
    "    elif use_rep in adata.layers.keys():\n",
    "        x = adata.layers[use_rep]\n",
    "        features = adata.var_names.values\n",
    "        if highly_variable:\n",
    "            x = x[:, k_hv]\n",
    "            features = features[k_hv]\n",
    "    elif use_rep in adata.obsm.keys():\n",
    "        x = adata.obsm[use_rep]\n",
    "        features = np.arange(x.shape[1])\n",
    "    elif (isinstance(use_rep, np.ndarray) and\n",
    "            use_rep.shape[0] == adata.shape[0]):\n",
    "        x = use_rep\n",
    "        features = np.arange(x.shape[1])\n",
    "    else:\n",
    "        raise KeyError(f'{use_rep} invalid.')\n",
    "    summarised = np.zeros((n_level, x.shape[1]))\n",
    "    for i, grp in enumerate(groups):\n",
    "        k_grp = group_attr == grp\n",
    "        if sp.issparse(x):\n",
    "            summarised[i] = FUN(x[k_grp, :], axis=0)\n",
    "        else:\n",
    "            summarised[i] = FUN(x[k_grp, :], axis=0, keepdims=True)\n",
    "    return pd.DataFrame(summarised.T, columns=groups, index=features)\n",
    "\n",
    "def random_partition(\n",
    "    adata,\n",
    "    partition_size,\n",
    "    groupby=None,\n",
    "    method=\"random_even\",\n",
    "    key_added=\"partition_labels\",\n",
    "    random_state=0,\n",
    "):\n",
    "    np.random.seed(random_state)\n",
    "    if groupby:\n",
    "        if groupby not in adata.obs.columns:\n",
    "            raise KeyError(f\"{groupby} is not a valid obs annotation.\")\n",
    "        groups = adata.obs[groupby].unique()\n",
    "        label_df = adata.obs[[groupby]].astype(str).rename(columns={groupby: key_added})\n",
    "        for grp in groups:\n",
    "            k = adata.obs[groupby] == grp\n",
    "            grp_size = sum(k)\n",
    "            n_partition = max(np.round(grp_size / partition_size).astype(int), 1)\n",
    "            if method == \"random\":\n",
    "                part_idx = np.random.randint(low=0, high=n_partition, size=grp_size)\n",
    "            elif method == \"random_even\":\n",
    "                part_sizes = list(map(len, np.array_split(np.arange(grp_size), n_partition)))\n",
    "                part_idx = np.repeat(np.arange(n_partition), part_sizes)\n",
    "                np.random.shuffle(part_idx)\n",
    "            else:\n",
    "                raise NotImplementedError(method)\n",
    "            label_df.loc[k, key_added] = [f\"{grp},{i}\" for i in part_idx]\n",
    "        adata.obs[key_added] = label_df[key_added]\n",
    "    else:\n",
    "        n_partition = max(np.round(adata.n_obs / partition_size).astype(int), 1)\n",
    "        if method == \"random\":\n",
    "            part_idx = np.random.randint(low=0, high=n_partition, size=adata.n_obs)\n",
    "        elif method == \"random_even\":\n",
    "            part_sizes = list(map(len, np.array_split(np.arange(adata.n_obs), n_partition)))\n",
    "            part_idx = np.repeat(np.arange(n_partition), part_sizes)\n",
    "            np.random.shuffle(part_idx)\n",
    "        else:\n",
    "            raise NotImplementedError(method)\n",
    "        adata.obs[key_added] = part_idx.astype(str)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "trained-spelling",
   "metadata": {},
   "source": [
    "## Import suspension data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "inclusive-victory",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 176050 × 3311\n",
       "    obs: 'latent_RT_efficiency', 'latent_cell_probability', 'latent_scale', 'sangerID', 'combinedID', 'donor', 'donor_type', 'region_finest', 'facility', 'cell_or_nuclei', 'modality', 'kit_10x', 'flushed', 'scrublet_score', 'scrublet_leiden', 'cluster_scrublet_score', 'doublet_pval', 'doublet_bh_pval', 'n_genes', 'n_counts', 'n_genes_by_counts', 'log1p_n_genes_by_counts', 'total_counts', 'log1p_total_counts', 'total_counts_mt', 'log1p_total_counts_mt', 'pct_counts_mt', 'total_counts_ribo', 'log1p_total_counts_ribo', 'pct_counts_ribo', 'HB_score', 'batch_key', '_scvi_batch', '_scvi_labels', 'annotation_Semih', 'leiden_scVI', 'cell_type_pre', 'pth', 'ito', 'FACSgate', 'leiden_scVI_0.05', 'leiden_scVI_0.01', 'leiden_scVI_0.1', 'leiden_scVI_0.15', 'leiden_scVI_0.2', 'leiden_scVI_0.25', 'leiden_scVI_0.3', 'leiden_scVI_0.35', 'leiden_scVI_0.4', 'cell_type', 'coarse_grain', 'fine_grain', 'region', 'sex', 'age', 'cycling', 'S_score', 'G2M_score', 'phase', 'stress_score', 'hb1_score'\n",
       "    var: 'gene_name_scRNA-0', 'gene_id', 'highly_variable', 'means', 'dispersions', 'dispersions_norm', 'highly_variable_nbatches', 'highly_variable_intersection'\n",
       "    uns: '_scvi_manager_uuid', '_scvi_uuid', 'age_colors', 'cell_or_nuclei_colors', 'cell_type_colors', 'coarse_grain_colors', 'cycling_colors', 'donor_colors', 'fine_grain_colors', 'hvg', 'kit_10x_colors', 'neighbors', 'phase_colors', 'region_colors', 'sex_colors', 'umap'\n",
       "    obsm: 'X_scVI', 'X_umap', '_scvi_extra_categorical_covs', '_scvi_extra_continuous', '_scvi_extra_continuous_covs', 'latent_gene_encoding'\n",
       "    layers: 'counts'\n",
       "    obsp: 'connectivities', 'distances'"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata = sc.read('/nfs/team205/heart/anndata_objects/Foetal/final_annotation/FetalHearts13Aug.h5ad')\n",
    "adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "rough-theorem",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(176050, 3311)\n",
      "(176050, 36601)\n"
     ]
    }
   ],
   "source": [
    "# Recover log-normed data for all genes to adata.X (allows Compass to calculate reactions across all genes, not just HVG-subsetted!)\n",
    "print(adata.shape)\n",
    "adata=adata.raw.to_adata()\n",
    "print(adata.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "mexican-catalyst",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Index(['MIR1302-2HG', 'FAM138A'], dtype='object', name='gene_name_multiome-1')\n",
      "[0.09802545 0.09802545 0.41538525 0.41538525 0.6009918 ]\n"
     ]
    }
   ],
   "source": [
    "# Check var_names are gene symbols, and the data is log-normed\n",
    "print(adata.var.index[:2])\n",
    "print(adata.X.data[:5])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dried-financing",
   "metadata": {},
   "source": [
    "# Subset to CMs only"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "clinical-chicken",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(176050, 36601)\n",
      "(64751, 36601)\n"
     ]
    }
   ],
   "source": [
    "print(adata.shape)\n",
    "adata=adata[adata.obs['coarse_grain']=='Cardiomyocytes']\n",
    "print(adata.shape)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "significant-bowling",
   "metadata": {},
   "source": [
    "# Prepare data for Compass\n",
    "\"The input gene expression matrix can be either a tab-delimited text file (tsv) or a matrix market format (mtx) containing gene expression estimates (CPM, TPM, or similar scaled units) with one row per gene, one column per sample.\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "touched-spank",
   "metadata": {},
   "source": [
    "## .mtx method (gene x sampleID, 3 files)\n",
    "This allows you to specify the 'samples', so rather than calculating reactions in all single cells individually (or micro-pooled) we can make our own \"samples\" (pseudobulking them)\n",
    "\n",
    "Need to ensure the samples are representative of the sample. Here we use a new batch key of cell state and the previous batch key  to form the 'samples'\n",
    "This requires 3 files, to inputted thus ```--data expression.mtx genes.tsv sample_names.tsv```\n",
    "\n",
    "Another option would be to use Milo or Metacell to make small samples, below I am going to just use the batch key to make samples within each cell state."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "rubber-clinton",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "VentricularCardiomyocytesCompact         13740\n",
       "VentricularCardiomyocytesPRRX1pos        11627\n",
       "VentricularCardiomyocytesCycling         11297\n",
       "VentricularCardiomyocytesTrabeculated     6936\n",
       "AtrialCardiomyocytesLeft                  6304\n",
       "AtrialCardiomyocytesRight                 5452\n",
       "Cardiofibromyocytes                       5278\n",
       "VentricularConductionSystemDistal         1568\n",
       "SinoatrialNodeCardiomyocytes               814\n",
       "VentricularConductionSystemProximal        730\n",
       "AtrioventricularNodeCardiomyocytes         689\n",
       "AtrialCardiomyocytesCycling                316\n",
       "Name: fine_grain, dtype: int64"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# make random groups of size 20 within each cell state\n",
    "adata.obs['fine_grain'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "martial-spray",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_792/3664867463.py:87: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  adata.obs[key_added] = label_df[key_added]\n"
     ]
    }
   ],
   "source": [
    "# Make random partition using sctk\n",
    "random_partition(adata, partition_size=10, groupby=\"fine_grain\", key_added=\"partition_10\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "sticky-compilation",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>latent_RT_efficiency</th>\n",
       "      <th>latent_cell_probability</th>\n",
       "      <th>latent_scale</th>\n",
       "      <th>sangerID</th>\n",
       "      <th>combinedID</th>\n",
       "      <th>donor</th>\n",
       "      <th>donor_type</th>\n",
       "      <th>region_finest</th>\n",
       "      <th>facility</th>\n",
       "      <th>cell_or_nuclei</th>\n",
       "      <th>...</th>\n",
       "      <th>region</th>\n",
       "      <th>sex</th>\n",
       "      <th>age</th>\n",
       "      <th>cycling</th>\n",
       "      <th>S_score</th>\n",
       "      <th>G2M_score</th>\n",
       "      <th>phase</th>\n",
       "      <th>stress_score</th>\n",
       "      <th>hb1_score</th>\n",
       "      <th>partition_10</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>barcode</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>BHF_F_Hea10402917_GATCGATAGACCCACC-1</th>\n",
       "      <td>3.064704</td>\n",
       "      <td>0.999994</td>\n",
       "      <td>10902.040039</td>\n",
       "      <td>BHF_F_Hea10402917</td>\n",
       "      <td>na</td>\n",
       "      <td>C86</td>\n",
       "      <td>na</td>\n",
       "      <td>Heart</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>Cell</td>\n",
       "      <td>...</td>\n",
       "      <td>WholeHeart</td>\n",
       "      <td>Male</td>\n",
       "      <td>12W5D</td>\n",
       "      <td>No</td>\n",
       "      <td>-0.057931</td>\n",
       "      <td>-0.131408</td>\n",
       "      <td>G1</td>\n",
       "      <td>0.640494</td>\n",
       "      <td>-0.196018</td>\n",
       "      <td>VentricularCardiomyocytesPRRX1pos,965</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BHF_F_Hea10402917_CATCAAGCATGGGACA-1</th>\n",
       "      <td>3.019039</td>\n",
       "      <td>0.999993</td>\n",
       "      <td>9811.644531</td>\n",
       "      <td>BHF_F_Hea10402917</td>\n",
       "      <td>na</td>\n",
       "      <td>C86</td>\n",
       "      <td>na</td>\n",
       "      <td>Heart</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>Cell</td>\n",
       "      <td>...</td>\n",
       "      <td>WholeHeart</td>\n",
       "      <td>Male</td>\n",
       "      <td>12W5D</td>\n",
       "      <td>Yes</td>\n",
       "      <td>1.306764</td>\n",
       "      <td>0.343963</td>\n",
       "      <td>S</td>\n",
       "      <td>-0.167570</td>\n",
       "      <td>-0.169170</td>\n",
       "      <td>VentricularCardiomyocytesCycling,900</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BHF_F_Hea10402917_CACCACTAGAGTGAGA-1</th>\n",
       "      <td>2.600170</td>\n",
       "      <td>0.999989</td>\n",
       "      <td>8616.544922</td>\n",
       "      <td>BHF_F_Hea10402917</td>\n",
       "      <td>na</td>\n",
       "      <td>C86</td>\n",
       "      <td>na</td>\n",
       "      <td>Heart</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>Cell</td>\n",
       "      <td>...</td>\n",
       "      <td>WholeHeart</td>\n",
       "      <td>Male</td>\n",
       "      <td>12W5D</td>\n",
       "      <td>No</td>\n",
       "      <td>-0.029281</td>\n",
       "      <td>-0.075899</td>\n",
       "      <td>G1</td>\n",
       "      <td>-0.012796</td>\n",
       "      <td>-0.177667</td>\n",
       "      <td>VentricularCardiomyocytesCompact,955</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BHF_F_Hea10402917_TCTTTCCTCTGGTGTA-1</th>\n",
       "      <td>2.481375</td>\n",
       "      <td>0.999992</td>\n",
       "      <td>8874.579102</td>\n",
       "      <td>BHF_F_Hea10402917</td>\n",
       "      <td>na</td>\n",
       "      <td>C86</td>\n",
       "      <td>na</td>\n",
       "      <td>Heart</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>Cell</td>\n",
       "      <td>...</td>\n",
       "      <td>WholeHeart</td>\n",
       "      <td>Male</td>\n",
       "      <td>12W5D</td>\n",
       "      <td>No</td>\n",
       "      <td>-0.015760</td>\n",
       "      <td>-0.130530</td>\n",
       "      <td>G1</td>\n",
       "      <td>-0.192063</td>\n",
       "      <td>-0.178246</td>\n",
       "      <td>VentricularCardiomyocytesTrabeculated,176</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BHF_F_Hea10402917_AAATGCCGTCAACATC-1</th>\n",
       "      <td>2.400682</td>\n",
       "      <td>0.999959</td>\n",
       "      <td>7570.651855</td>\n",
       "      <td>BHF_F_Hea10402917</td>\n",
       "      <td>na</td>\n",
       "      <td>C86</td>\n",
       "      <td>na</td>\n",
       "      <td>Heart</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>Cell</td>\n",
       "      <td>...</td>\n",
       "      <td>WholeHeart</td>\n",
       "      <td>Male</td>\n",
       "      <td>12W5D</td>\n",
       "      <td>No</td>\n",
       "      <td>-0.198699</td>\n",
       "      <td>-0.225604</td>\n",
       "      <td>G1</td>\n",
       "      <td>1.522051</td>\n",
       "      <td>-0.289346</td>\n",
       "      <td>Cardiofibromyocytes,18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BHF_F_Hea11933675_BHF_F_Hea11596628_CCTGTTGGTTGTGACA-1</th>\n",
       "      <td>0.977603</td>\n",
       "      <td>0.998388</td>\n",
       "      <td>2734.687744</td>\n",
       "      <td>BHF_F_Hea11933675</td>\n",
       "      <td>BHF_F_Hea11933675_BHF_F_Hea11596628</td>\n",
       "      <td>C98</td>\n",
       "      <td>na</td>\n",
       "      <td>Heart</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>Nuclei</td>\n",
       "      <td>...</td>\n",
       "      <td>WholeHeart</td>\n",
       "      <td>Male</td>\n",
       "      <td>9W1D</td>\n",
       "      <td>No</td>\n",
       "      <td>-0.071211</td>\n",
       "      <td>-0.173691</td>\n",
       "      <td>G1</td>\n",
       "      <td>-0.367560</td>\n",
       "      <td>0.247539</td>\n",
       "      <td>VentricularCardiomyocytesPRRX1pos,1079</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BHF_F_Hea11933675_BHF_F_Hea11596628_GCGCTAGGTGTTGCTT-1</th>\n",
       "      <td>0.861791</td>\n",
       "      <td>0.997482</td>\n",
       "      <td>4039.140381</td>\n",
       "      <td>BHF_F_Hea11933675</td>\n",
       "      <td>BHF_F_Hea11933675_BHF_F_Hea11596628</td>\n",
       "      <td>C98</td>\n",
       "      <td>na</td>\n",
       "      <td>Heart</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>Nuclei</td>\n",
       "      <td>...</td>\n",
       "      <td>WholeHeart</td>\n",
       "      <td>Male</td>\n",
       "      <td>9W1D</td>\n",
       "      <td>No</td>\n",
       "      <td>0.054354</td>\n",
       "      <td>-0.197448</td>\n",
       "      <td>S</td>\n",
       "      <td>-0.389992</td>\n",
       "      <td>0.570969</td>\n",
       "      <td>Cardiofibromyocytes,157</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BHF_F_Hea11933675_BHF_F_Hea11596628_AGCTACGTCTTGTCTG-1</th>\n",
       "      <td>0.811526</td>\n",
       "      <td>0.990161</td>\n",
       "      <td>4187.259766</td>\n",
       "      <td>BHF_F_Hea11933675</td>\n",
       "      <td>BHF_F_Hea11933675_BHF_F_Hea11596628</td>\n",
       "      <td>C98</td>\n",
       "      <td>na</td>\n",
       "      <td>Heart</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>Nuclei</td>\n",
       "      <td>...</td>\n",
       "      <td>WholeHeart</td>\n",
       "      <td>Male</td>\n",
       "      <td>9W1D</td>\n",
       "      <td>Yes</td>\n",
       "      <td>-0.081391</td>\n",
       "      <td>-0.079086</td>\n",
       "      <td>G1</td>\n",
       "      <td>0.232271</td>\n",
       "      <td>0.579496</td>\n",
       "      <td>VentricularCardiomyocytesCycling,468</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BHF_F_Hea11933675_BHF_F_Hea11596628_CTGCTATGTTTGGTTC-1</th>\n",
       "      <td>0.807019</td>\n",
       "      <td>0.987438</td>\n",
       "      <td>3982.637451</td>\n",
       "      <td>BHF_F_Hea11933675</td>\n",
       "      <td>BHF_F_Hea11933675_BHF_F_Hea11596628</td>\n",
       "      <td>C98</td>\n",
       "      <td>na</td>\n",
       "      <td>Heart</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>Nuclei</td>\n",
       "      <td>...</td>\n",
       "      <td>WholeHeart</td>\n",
       "      <td>Male</td>\n",
       "      <td>9W1D</td>\n",
       "      <td>Yes</td>\n",
       "      <td>-0.070203</td>\n",
       "      <td>-0.121220</td>\n",
       "      <td>G1</td>\n",
       "      <td>-0.365374</td>\n",
       "      <td>0.868231</td>\n",
       "      <td>VentricularCardiomyocytesPRRX1pos,273</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BHF_F_Hea11933675_BHF_F_Hea11596628_GGCCTTAAGGCCTAAT-1</th>\n",
       "      <td>0.769857</td>\n",
       "      <td>0.965501</td>\n",
       "      <td>3725.604004</td>\n",
       "      <td>BHF_F_Hea11933675</td>\n",
       "      <td>BHF_F_Hea11933675_BHF_F_Hea11596628</td>\n",
       "      <td>C98</td>\n",
       "      <td>na</td>\n",
       "      <td>Heart</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>Nuclei</td>\n",
       "      <td>...</td>\n",
       "      <td>WholeHeart</td>\n",
       "      <td>Male</td>\n",
       "      <td>9W1D</td>\n",
       "      <td>Yes</td>\n",
       "      <td>0.115479</td>\n",
       "      <td>-0.161365</td>\n",
       "      <td>S</td>\n",
       "      <td>-0.015999</td>\n",
       "      <td>0.371164</td>\n",
       "      <td>Cardiofibromyocytes,11</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>64751 rows × 62 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                    latent_RT_efficiency  \\\n",
       "barcode                                                                    \n",
       "BHF_F_Hea10402917_GATCGATAGACCCACC-1                            3.064704   \n",
       "BHF_F_Hea10402917_CATCAAGCATGGGACA-1                            3.019039   \n",
       "BHF_F_Hea10402917_CACCACTAGAGTGAGA-1                            2.600170   \n",
       "BHF_F_Hea10402917_TCTTTCCTCTGGTGTA-1                            2.481375   \n",
       "BHF_F_Hea10402917_AAATGCCGTCAACATC-1                            2.400682   \n",
       "...                                                                  ...   \n",
       "BHF_F_Hea11933675_BHF_F_Hea11596628_CCTGTTGGTTG...              0.977603   \n",
       "BHF_F_Hea11933675_BHF_F_Hea11596628_GCGCTAGGTGT...              0.861791   \n",
       "BHF_F_Hea11933675_BHF_F_Hea11596628_AGCTACGTCTT...              0.811526   \n",
       "BHF_F_Hea11933675_BHF_F_Hea11596628_CTGCTATGTTT...              0.807019   \n",
       "BHF_F_Hea11933675_BHF_F_Hea11596628_GGCCTTAAGGC...              0.769857   \n",
       "\n",
       "                                                    latent_cell_probability  \\\n",
       "barcode                                                                       \n",
       "BHF_F_Hea10402917_GATCGATAGACCCACC-1                               0.999994   \n",
       "BHF_F_Hea10402917_CATCAAGCATGGGACA-1                               0.999993   \n",
       "BHF_F_Hea10402917_CACCACTAGAGTGAGA-1                               0.999989   \n",
       "BHF_F_Hea10402917_TCTTTCCTCTGGTGTA-1                               0.999992   \n",
       "BHF_F_Hea10402917_AAATGCCGTCAACATC-1                               0.999959   \n",
       "...                                                                     ...   \n",
       "BHF_F_Hea11933675_BHF_F_Hea11596628_CCTGTTGGTTG...                 0.998388   \n",
       "BHF_F_Hea11933675_BHF_F_Hea11596628_GCGCTAGGTGT...                 0.997482   \n",
       "BHF_F_Hea11933675_BHF_F_Hea11596628_AGCTACGTCTT...                 0.990161   \n",
       "BHF_F_Hea11933675_BHF_F_Hea11596628_CTGCTATGTTT...                 0.987438   \n",
       "BHF_F_Hea11933675_BHF_F_Hea11596628_GGCCTTAAGGC...                 0.965501   \n",
       "\n",
       "                                                    latent_scale  \\\n",
       "barcode                                                            \n",
       "BHF_F_Hea10402917_GATCGATAGACCCACC-1                10902.040039   \n",
       "BHF_F_Hea10402917_CATCAAGCATGGGACA-1                 9811.644531   \n",
       "BHF_F_Hea10402917_CACCACTAGAGTGAGA-1                 8616.544922   \n",
       "BHF_F_Hea10402917_TCTTTCCTCTGGTGTA-1                 8874.579102   \n",
       "BHF_F_Hea10402917_AAATGCCGTCAACATC-1                 7570.651855   \n",
       "...                                                          ...   \n",
       "BHF_F_Hea11933675_BHF_F_Hea11596628_CCTGTTGGTTG...   2734.687744   \n",
       "BHF_F_Hea11933675_BHF_F_Hea11596628_GCGCTAGGTGT...   4039.140381   \n",
       "BHF_F_Hea11933675_BHF_F_Hea11596628_AGCTACGTCTT...   4187.259766   \n",
       "BHF_F_Hea11933675_BHF_F_Hea11596628_CTGCTATGTTT...   3982.637451   \n",
       "BHF_F_Hea11933675_BHF_F_Hea11596628_GGCCTTAAGGC...   3725.604004   \n",
       "\n",
       "                                                             sangerID  \\\n",
       "barcode                                                                 \n",
       "BHF_F_Hea10402917_GATCGATAGACCCACC-1                BHF_F_Hea10402917   \n",
       "BHF_F_Hea10402917_CATCAAGCATGGGACA-1                BHF_F_Hea10402917   \n",
       "BHF_F_Hea10402917_CACCACTAGAGTGAGA-1                BHF_F_Hea10402917   \n",
       "BHF_F_Hea10402917_TCTTTCCTCTGGTGTA-1                BHF_F_Hea10402917   \n",
       "BHF_F_Hea10402917_AAATGCCGTCAACATC-1                BHF_F_Hea10402917   \n",
       "...                                                               ...   \n",
       "BHF_F_Hea11933675_BHF_F_Hea11596628_CCTGTTGGTTG...  BHF_F_Hea11933675   \n",
       "BHF_F_Hea11933675_BHF_F_Hea11596628_GCGCTAGGTGT...  BHF_F_Hea11933675   \n",
       "BHF_F_Hea11933675_BHF_F_Hea11596628_AGCTACGTCTT...  BHF_F_Hea11933675   \n",
       "BHF_F_Hea11933675_BHF_F_Hea11596628_CTGCTATGTTT...  BHF_F_Hea11933675   \n",
       "BHF_F_Hea11933675_BHF_F_Hea11596628_GGCCTTAAGGC...  BHF_F_Hea11933675   \n",
       "\n",
       "                                                                             combinedID  \\\n",
       "barcode                                                                                   \n",
       "BHF_F_Hea10402917_GATCGATAGACCCACC-1                                                 na   \n",
       "BHF_F_Hea10402917_CATCAAGCATGGGACA-1                                                 na   \n",
       "BHF_F_Hea10402917_CACCACTAGAGTGAGA-1                                                 na   \n",
       "BHF_F_Hea10402917_TCTTTCCTCTGGTGTA-1                                                 na   \n",
       "BHF_F_Hea10402917_AAATGCCGTCAACATC-1                                                 na   \n",
       "...                                                                                 ...   \n",
       "BHF_F_Hea11933675_BHF_F_Hea11596628_CCTGTTGGTTG...  BHF_F_Hea11933675_BHF_F_Hea11596628   \n",
       "BHF_F_Hea11933675_BHF_F_Hea11596628_GCGCTAGGTGT...  BHF_F_Hea11933675_BHF_F_Hea11596628   \n",
       "BHF_F_Hea11933675_BHF_F_Hea11596628_AGCTACGTCTT...  BHF_F_Hea11933675_BHF_F_Hea11596628   \n",
       "BHF_F_Hea11933675_BHF_F_Hea11596628_CTGCTATGTTT...  BHF_F_Hea11933675_BHF_F_Hea11596628   \n",
       "BHF_F_Hea11933675_BHF_F_Hea11596628_GGCCTTAAGGC...  BHF_F_Hea11933675_BHF_F_Hea11596628   \n",
       "\n",
       "                                                   donor donor_type  \\\n",
       "barcode                                                               \n",
       "BHF_F_Hea10402917_GATCGATAGACCCACC-1                 C86         na   \n",
       "BHF_F_Hea10402917_CATCAAGCATGGGACA-1                 C86         na   \n",
       "BHF_F_Hea10402917_CACCACTAGAGTGAGA-1                 C86         na   \n",
       "BHF_F_Hea10402917_TCTTTCCTCTGGTGTA-1                 C86         na   \n",
       "BHF_F_Hea10402917_AAATGCCGTCAACATC-1                 C86         na   \n",
       "...                                                  ...        ...   \n",
       "BHF_F_Hea11933675_BHF_F_Hea11596628_CCTGTTGGTTG...   C98         na   \n",
       "BHF_F_Hea11933675_BHF_F_Hea11596628_GCGCTAGGTGT...   C98         na   \n",
       "BHF_F_Hea11933675_BHF_F_Hea11596628_AGCTACGTCTT...   C98         na   \n",
       "BHF_F_Hea11933675_BHF_F_Hea11596628_CTGCTATGTTT...   C98         na   \n",
       "BHF_F_Hea11933675_BHF_F_Hea11596628_GGCCTTAAGGC...   C98         na   \n",
       "\n",
       "                                                   region_finest facility  \\\n",
       "barcode                                                                     \n",
       "BHF_F_Hea10402917_GATCGATAGACCCACC-1                       Heart   Sanger   \n",
       "BHF_F_Hea10402917_CATCAAGCATGGGACA-1                       Heart   Sanger   \n",
       "BHF_F_Hea10402917_CACCACTAGAGTGAGA-1                       Heart   Sanger   \n",
       "BHF_F_Hea10402917_TCTTTCCTCTGGTGTA-1                       Heart   Sanger   \n",
       "BHF_F_Hea10402917_AAATGCCGTCAACATC-1                       Heart   Sanger   \n",
       "...                                                          ...      ...   \n",
       "BHF_F_Hea11933675_BHF_F_Hea11596628_CCTGTTGGTTG...         Heart   Sanger   \n",
       "BHF_F_Hea11933675_BHF_F_Hea11596628_GCGCTAGGTGT...         Heart   Sanger   \n",
       "BHF_F_Hea11933675_BHF_F_Hea11596628_AGCTACGTCTT...         Heart   Sanger   \n",
       "BHF_F_Hea11933675_BHF_F_Hea11596628_CTGCTATGTTT...         Heart   Sanger   \n",
       "BHF_F_Hea11933675_BHF_F_Hea11596628_GGCCTTAAGGC...         Heart   Sanger   \n",
       "\n",
       "                                                   cell_or_nuclei  ...  \\\n",
       "barcode                                                            ...   \n",
       "BHF_F_Hea10402917_GATCGATAGACCCACC-1                         Cell  ...   \n",
       "BHF_F_Hea10402917_CATCAAGCATGGGACA-1                         Cell  ...   \n",
       "BHF_F_Hea10402917_CACCACTAGAGTGAGA-1                         Cell  ...   \n",
       "BHF_F_Hea10402917_TCTTTCCTCTGGTGTA-1                         Cell  ...   \n",
       "BHF_F_Hea10402917_AAATGCCGTCAACATC-1                         Cell  ...   \n",
       "...                                                           ...  ...   \n",
       "BHF_F_Hea11933675_BHF_F_Hea11596628_CCTGTTGGTTG...         Nuclei  ...   \n",
       "BHF_F_Hea11933675_BHF_F_Hea11596628_GCGCTAGGTGT...         Nuclei  ...   \n",
       "BHF_F_Hea11933675_BHF_F_Hea11596628_AGCTACGTCTT...         Nuclei  ...   \n",
       "BHF_F_Hea11933675_BHF_F_Hea11596628_CTGCTATGTTT...         Nuclei  ...   \n",
       "BHF_F_Hea11933675_BHF_F_Hea11596628_GGCCTTAAGGC...         Nuclei  ...   \n",
       "\n",
       "                                                        region   sex    age  \\\n",
       "barcode                                                                       \n",
       "BHF_F_Hea10402917_GATCGATAGACCCACC-1                WholeHeart  Male  12W5D   \n",
       "BHF_F_Hea10402917_CATCAAGCATGGGACA-1                WholeHeart  Male  12W5D   \n",
       "BHF_F_Hea10402917_CACCACTAGAGTGAGA-1                WholeHeart  Male  12W5D   \n",
       "BHF_F_Hea10402917_TCTTTCCTCTGGTGTA-1                WholeHeart  Male  12W5D   \n",
       "BHF_F_Hea10402917_AAATGCCGTCAACATC-1                WholeHeart  Male  12W5D   \n",
       "...                                                        ...   ...    ...   \n",
       "BHF_F_Hea11933675_BHF_F_Hea11596628_CCTGTTGGTTG...  WholeHeart  Male   9W1D   \n",
       "BHF_F_Hea11933675_BHF_F_Hea11596628_GCGCTAGGTGT...  WholeHeart  Male   9W1D   \n",
       "BHF_F_Hea11933675_BHF_F_Hea11596628_AGCTACGTCTT...  WholeHeart  Male   9W1D   \n",
       "BHF_F_Hea11933675_BHF_F_Hea11596628_CTGCTATGTTT...  WholeHeart  Male   9W1D   \n",
       "BHF_F_Hea11933675_BHF_F_Hea11596628_GGCCTTAAGGC...  WholeHeart  Male   9W1D   \n",
       "\n",
       "                                                    cycling   S_score  \\\n",
       "barcode                                                                 \n",
       "BHF_F_Hea10402917_GATCGATAGACCCACC-1                     No -0.057931   \n",
       "BHF_F_Hea10402917_CATCAAGCATGGGACA-1                    Yes  1.306764   \n",
       "BHF_F_Hea10402917_CACCACTAGAGTGAGA-1                     No -0.029281   \n",
       "BHF_F_Hea10402917_TCTTTCCTCTGGTGTA-1                     No -0.015760   \n",
       "BHF_F_Hea10402917_AAATGCCGTCAACATC-1                     No -0.198699   \n",
       "...                                                     ...       ...   \n",
       "BHF_F_Hea11933675_BHF_F_Hea11596628_CCTGTTGGTTG...       No -0.071211   \n",
       "BHF_F_Hea11933675_BHF_F_Hea11596628_GCGCTAGGTGT...       No  0.054354   \n",
       "BHF_F_Hea11933675_BHF_F_Hea11596628_AGCTACGTCTT...      Yes -0.081391   \n",
       "BHF_F_Hea11933675_BHF_F_Hea11596628_CTGCTATGTTT...      Yes -0.070203   \n",
       "BHF_F_Hea11933675_BHF_F_Hea11596628_GGCCTTAAGGC...      Yes  0.115479   \n",
       "\n",
       "                                                    G2M_score  phase  \\\n",
       "barcode                                                                \n",
       "BHF_F_Hea10402917_GATCGATAGACCCACC-1                -0.131408     G1   \n",
       "BHF_F_Hea10402917_CATCAAGCATGGGACA-1                 0.343963      S   \n",
       "BHF_F_Hea10402917_CACCACTAGAGTGAGA-1                -0.075899     G1   \n",
       "BHF_F_Hea10402917_TCTTTCCTCTGGTGTA-1                -0.130530     G1   \n",
       "BHF_F_Hea10402917_AAATGCCGTCAACATC-1                -0.225604     G1   \n",
       "...                                                       ...    ...   \n",
       "BHF_F_Hea11933675_BHF_F_Hea11596628_CCTGTTGGTTG...  -0.173691     G1   \n",
       "BHF_F_Hea11933675_BHF_F_Hea11596628_GCGCTAGGTGT...  -0.197448      S   \n",
       "BHF_F_Hea11933675_BHF_F_Hea11596628_AGCTACGTCTT...  -0.079086     G1   \n",
       "BHF_F_Hea11933675_BHF_F_Hea11596628_CTGCTATGTTT...  -0.121220     G1   \n",
       "BHF_F_Hea11933675_BHF_F_Hea11596628_GGCCTTAAGGC...  -0.161365      S   \n",
       "\n",
       "                                                    stress_score  hb1_score  \\\n",
       "barcode                                                                       \n",
       "BHF_F_Hea10402917_GATCGATAGACCCACC-1                    0.640494  -0.196018   \n",
       "BHF_F_Hea10402917_CATCAAGCATGGGACA-1                   -0.167570  -0.169170   \n",
       "BHF_F_Hea10402917_CACCACTAGAGTGAGA-1                   -0.012796  -0.177667   \n",
       "BHF_F_Hea10402917_TCTTTCCTCTGGTGTA-1                   -0.192063  -0.178246   \n",
       "BHF_F_Hea10402917_AAATGCCGTCAACATC-1                    1.522051  -0.289346   \n",
       "...                                                          ...        ...   \n",
       "BHF_F_Hea11933675_BHF_F_Hea11596628_CCTGTTGGTTG...     -0.367560   0.247539   \n",
       "BHF_F_Hea11933675_BHF_F_Hea11596628_GCGCTAGGTGT...     -0.389992   0.570969   \n",
       "BHF_F_Hea11933675_BHF_F_Hea11596628_AGCTACGTCTT...      0.232271   0.579496   \n",
       "BHF_F_Hea11933675_BHF_F_Hea11596628_CTGCTATGTTT...     -0.365374   0.868231   \n",
       "BHF_F_Hea11933675_BHF_F_Hea11596628_GGCCTTAAGGC...     -0.015999   0.371164   \n",
       "\n",
       "                                                                                 partition_10  \n",
       "barcode                                                                                        \n",
       "BHF_F_Hea10402917_GATCGATAGACCCACC-1                    VentricularCardiomyocytesPRRX1pos,965  \n",
       "BHF_F_Hea10402917_CATCAAGCATGGGACA-1                     VentricularCardiomyocytesCycling,900  \n",
       "BHF_F_Hea10402917_CACCACTAGAGTGAGA-1                     VentricularCardiomyocytesCompact,955  \n",
       "BHF_F_Hea10402917_TCTTTCCTCTGGTGTA-1                VentricularCardiomyocytesTrabeculated,176  \n",
       "BHF_F_Hea10402917_AAATGCCGTCAACATC-1                                   Cardiofibromyocytes,18  \n",
       "...                                                                                       ...  \n",
       "BHF_F_Hea11933675_BHF_F_Hea11596628_CCTGTTGGTTG...     VentricularCardiomyocytesPRRX1pos,1079  \n",
       "BHF_F_Hea11933675_BHF_F_Hea11596628_GCGCTAGGTGT...                    Cardiofibromyocytes,157  \n",
       "BHF_F_Hea11933675_BHF_F_Hea11596628_AGCTACGTCTT...       VentricularCardiomyocytesCycling,468  \n",
       "BHF_F_Hea11933675_BHF_F_Hea11596628_CTGCTATGTTT...      VentricularCardiomyocytesPRRX1pos,273  \n",
       "BHF_F_Hea11933675_BHF_F_Hea11596628_GGCCTTAAGGC...                     Cardiofibromyocytes,11  \n",
       "\n",
       "[64751 rows x 62 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.obs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "configured-solid",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Pseudobulk use Ni's psudobulk function\n",
    "pseudo_bulk = pseudo_bulk(adata,\n",
    "                         groupby='partition_10')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "missing-villa",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(36601, 6476)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>AtrialCardiomyocytesCycling,0</th>\n",
       "      <th>AtrialCardiomyocytesCycling,1</th>\n",
       "      <th>AtrialCardiomyocytesCycling,10</th>\n",
       "      <th>AtrialCardiomyocytesCycling,11</th>\n",
       "      <th>AtrialCardiomyocytesCycling,12</th>\n",
       "      <th>AtrialCardiomyocytesCycling,13</th>\n",
       "      <th>AtrialCardiomyocytesCycling,14</th>\n",
       "      <th>AtrialCardiomyocytesCycling,15</th>\n",
       "      <th>AtrialCardiomyocytesCycling,16</th>\n",
       "      <th>AtrialCardiomyocytesCycling,17</th>\n",
       "      <th>...</th>\n",
       "      <th>VentricularConductionSystemProximal,66</th>\n",
       "      <th>VentricularConductionSystemProximal,67</th>\n",
       "      <th>VentricularConductionSystemProximal,68</th>\n",
       "      <th>VentricularConductionSystemProximal,69</th>\n",
       "      <th>VentricularConductionSystemProximal,7</th>\n",
       "      <th>VentricularConductionSystemProximal,70</th>\n",
       "      <th>VentricularConductionSystemProximal,71</th>\n",
       "      <th>VentricularConductionSystemProximal,72</th>\n",
       "      <th>VentricularConductionSystemProximal,8</th>\n",
       "      <th>VentricularConductionSystemProximal,9</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>MIR1302-2HG</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FAM138A</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OR4F5</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 6476 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             AtrialCardiomyocytesCycling,0  AtrialCardiomyocytesCycling,1  \\\n",
       "MIR1302-2HG                            0.0                            0.0   \n",
       "FAM138A                                0.0                            0.0   \n",
       "OR4F5                                  0.0                            0.0   \n",
       "\n",
       "             AtrialCardiomyocytesCycling,10  AtrialCardiomyocytesCycling,11  \\\n",
       "MIR1302-2HG                             0.0                             0.0   \n",
       "FAM138A                                 0.0                             0.0   \n",
       "OR4F5                                   0.0                             0.0   \n",
       "\n",
       "             AtrialCardiomyocytesCycling,12  AtrialCardiomyocytesCycling,13  \\\n",
       "MIR1302-2HG                             0.0                             0.0   \n",
       "FAM138A                                 0.0                             0.0   \n",
       "OR4F5                                   0.0                             0.0   \n",
       "\n",
       "             AtrialCardiomyocytesCycling,14  AtrialCardiomyocytesCycling,15  \\\n",
       "MIR1302-2HG                             0.0                             0.0   \n",
       "FAM138A                                 0.0                             0.0   \n",
       "OR4F5                                   0.0                             0.0   \n",
       "\n",
       "             AtrialCardiomyocytesCycling,16  AtrialCardiomyocytesCycling,17  \\\n",
       "MIR1302-2HG                             0.0                             0.0   \n",
       "FAM138A                                 0.0                             0.0   \n",
       "OR4F5                                   0.0                             0.0   \n",
       "\n",
       "             ...  VentricularConductionSystemProximal,66  \\\n",
       "MIR1302-2HG  ...                                     0.0   \n",
       "FAM138A      ...                                     0.0   \n",
       "OR4F5        ...                                     0.0   \n",
       "\n",
       "             VentricularConductionSystemProximal,67  \\\n",
       "MIR1302-2HG                                     0.0   \n",
       "FAM138A                                         0.0   \n",
       "OR4F5                                           0.0   \n",
       "\n",
       "             VentricularConductionSystemProximal,68  \\\n",
       "MIR1302-2HG                                     0.0   \n",
       "FAM138A                                         0.0   \n",
       "OR4F5                                           0.0   \n",
       "\n",
       "             VentricularConductionSystemProximal,69  \\\n",
       "MIR1302-2HG                                     0.0   \n",
       "FAM138A                                         0.0   \n",
       "OR4F5                                           0.0   \n",
       "\n",
       "             VentricularConductionSystemProximal,7  \\\n",
       "MIR1302-2HG                                    0.0   \n",
       "FAM138A                                        0.0   \n",
       "OR4F5                                          0.0   \n",
       "\n",
       "             VentricularConductionSystemProximal,70  \\\n",
       "MIR1302-2HG                                     0.0   \n",
       "FAM138A                                         0.0   \n",
       "OR4F5                                           0.0   \n",
       "\n",
       "             VentricularConductionSystemProximal,71  \\\n",
       "MIR1302-2HG                                     0.0   \n",
       "FAM138A                                         0.0   \n",
       "OR4F5                                           0.0   \n",
       "\n",
       "             VentricularConductionSystemProximal,72  \\\n",
       "MIR1302-2HG                                     0.0   \n",
       "FAM138A                                         0.0   \n",
       "OR4F5                                           0.0   \n",
       "\n",
       "             VentricularConductionSystemProximal,8  \\\n",
       "MIR1302-2HG                                    0.0   \n",
       "FAM138A                                        0.0   \n",
       "OR4F5                                          0.0   \n",
       "\n",
       "             VentricularConductionSystemProximal,9  \n",
       "MIR1302-2HG                                    0.0  \n",
       "FAM138A                                        0.0  \n",
       "OR4F5                                          0.0  \n",
       "\n",
       "[3 rows x 6476 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(pseudo_bulk.shape)\n",
    "pseudo_bulk.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "digital-acting",
   "metadata": {},
   "outputs": [],
   "source": [
    "# make sample_names.tsv\n",
    "pseudo_bulk.columns.to_frame().to_csv('/nfs/team205/heart/compass/fetal_allCM_random10/sample_names.tsv', sep=\"\\t\", index=False, header=False)\n",
    "\n",
    "\n",
    "# make genes.tsv\n",
    "genes = pseudo_bulk.index.to_list()\n",
    "genes = pd.DataFrame(genes)\n",
    "genes.to_csv('/nfs/team205/heart/compass/fetal_allCM_random10/genes.tsv', sep=\"\\t\", index = False, header=False)\n",
    "\n",
    "# make expression.mtx\n",
    "from scipy.io import mmwrite, mmread\n",
    "from scipy.sparse import csr_matrix\n",
    "\n",
    "# Count Matrix (log normed, all genes, not HVG subsetted)\n",
    "sparce_mtx = csr_matrix(pseudo_bulk)\n",
    "sparce_mtx = sparce_mtx\n",
    "mmwrite('/nfs/team205/heart/compass/fetal_allCM_random10/expression.mtx', sparce_mtx)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "turned-occurrence",
   "metadata": {},
   "source": [
    "# Export cell metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "assured-superintendent",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Alter so it includes things you need (a reference e.g. barcode and a cell type e.g. in fine_grain)\n",
    "adata.obs.to_csv('/nfs/team205/heart/compass/fetal_allCM_random10/cell_metadata.csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "lucky-vegetarian",
   "metadata": {},
   "source": [
    "# Run Compass"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "extraordinary-phone",
   "metadata": {},
   "source": [
    "## bsub command\n",
    "\n",
    "```\n",
    "bsub \\\n",
    "-G teichlab \\\n",
    "-q basement \\\n",
    "-M300000 \\\n",
    "-R \"select[mem>300000] rusage[mem=300000] span[hosts=1]\" \\\n",
    "-n 30 \\\n",
    "-J all_pool50 \\\n",
    "-o 'log/compass.%J.out' \\\n",
    "-e 'log/compass.%J.err' \\\n",
    "'/software/singularity-v3.6.4/bin/singularity run -B /nfs,/lustre -B ~/COMPASSResources:/usr/local/lib/python3.8/site-packages/compass/Resources /nfs/cellgeni/singularity/images/compass-v0.9.10.2-cplex-v20.10.sif compass --data /nfs/team205/heart/compass/fetal_allcells/fetal_allcells_count_matrix.tsv --species homo_sapiens --precache --calc-metabolites --output-dir /nfs/team205/heart/compass/fetal_allcells/out/ --detailed-perf --num-processes 30 --microcluster-size 50'\n",
    "```\n",
    "\n",
    "remove --microcluster-size flag to run on single-cell data\n",
    "\n",
    "if using .mtx methods then need to change --data input e.g. \"--data expression.mtx genes.tsv sample_names.tsv\""
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "scanpy",
   "language": "python",
   "name": "scanpy"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
