{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "described-machinery",
   "metadata": {},
   "source": [
    "# Prepare Compass inputs, random groups within cell states"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "amber-worth",
   "metadata": {},
   "source": [
    "Due to large numbers of micropool clusters being contaminated with cells from other annotated groups, we will divide up cells within each celltype into groups of 20 (CM) and 50 (all cells)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "genuine-patio",
   "metadata": {},
   "source": [
    "## Setup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "suspended-debut",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/jovyan/my-conda-envs/scanpy/lib/python3.9/site-packages/tqdm/auto.py:22: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html\n",
      "  from .autonotebook import tqdm as notebook_tqdm\n"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import scanpy as sc\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "from matplotlib.pyplot import rc_context\n",
    "import bbknn\n",
    "import re\n",
    "import sctk"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "secure-football",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "scanpy==1.9.1 anndata==0.8.0 umap==0.5.3 numpy==1.22.4 scipy==1.8.1 pandas==1.4.3 scikit-learn==1.1.1 statsmodels==0.13.2 python-igraph==0.9.11 pynndescent==0.5.7\n"
     ]
    }
   ],
   "source": [
    "sc.settings.verbosity = 0            # verbosity: errors (0), warnings (1), info (2), hints (3)\n",
    "sc.logging.print_header()\n",
    "sc.settings.set_figure_params(dpi=80, facecolor='white', color_map = 'Reds', figsize = (5, 5),dpi_save=300)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "interstate-thomson",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Perform pseudobulk on sc data\n",
    "\n",
    "import scipy.sparse as sp\n",
    "\n",
    "def pseudo_bulk(\n",
    "        adata, groupby, use_rep='X', highly_variable=False, FUN=np.mean):\n",
    "    \"\"\"Make pseudo bulk data from grouped sc data\n",
    "    \"\"\"\n",
    "    if adata.obs[groupby].dtype.name == 'category':\n",
    "        group_attr = adata.obs[groupby].values\n",
    "        groups = adata.obs[groupby].cat.categories.values\n",
    "    else:\n",
    "        group_attr = adata.obs[groupby].astype(str).values\n",
    "        groups = np.unique(group_attr)\n",
    "    n_level = len(groups)\n",
    "    if highly_variable:\n",
    "        if isinstance(highly_variable, (list, tuple)):\n",
    "            if use_rep == 'raw':\n",
    "                k_hv = adata.raw.var_names.isin(highly_variable)\n",
    "            else:\n",
    "                k_hv = adata.var_names.isin(highly_variable)\n",
    "        else:\n",
    "            k_hv = adata.var['highly_variable'].values\n",
    "    if use_rep == 'X':\n",
    "        x = adata.X\n",
    "        features = adata.var_names.values\n",
    "        if highly_variable:\n",
    "            x = x[:, k_hv]\n",
    "            features = features[k_hv]\n",
    "    elif use_rep == 'raw':\n",
    "        x = adata.raw.X\n",
    "        features = adata.raw.var_names.values\n",
    "        if highly_variable:\n",
    "            x = x[:, k_hv]\n",
    "            features = features[k_hv]\n",
    "    elif use_rep in adata.layers.keys():\n",
    "        x = adata.layers[use_rep]\n",
    "        features = adata.var_names.values\n",
    "        if highly_variable:\n",
    "            x = x[:, k_hv]\n",
    "            features = features[k_hv]\n",
    "    elif use_rep in adata.obsm.keys():\n",
    "        x = adata.obsm[use_rep]\n",
    "        features = np.arange(x.shape[1])\n",
    "    elif (isinstance(use_rep, np.ndarray) and\n",
    "            use_rep.shape[0] == adata.shape[0]):\n",
    "        x = use_rep\n",
    "        features = np.arange(x.shape[1])\n",
    "    else:\n",
    "        raise KeyError(f'{use_rep} invalid.')\n",
    "    summarised = np.zeros((n_level, x.shape[1]))\n",
    "    for i, grp in enumerate(groups):\n",
    "        k_grp = group_attr == grp\n",
    "        if sp.issparse(x):\n",
    "            summarised[i] = FUN(x[k_grp, :], axis=0)\n",
    "        else:\n",
    "            summarised[i] = FUN(x[k_grp, :], axis=0, keepdims=True)\n",
    "    return pd.DataFrame(summarised.T, columns=groups, index=features)\n",
    "\n",
    "def random_partition(\n",
    "    adata,\n",
    "    partition_size,\n",
    "    groupby=None,\n",
    "    method=\"random_even\",\n",
    "    key_added=\"partition_labels\",\n",
    "    random_state=0,\n",
    "):\n",
    "    np.random.seed(random_state)\n",
    "    if groupby:\n",
    "        if groupby not in adata.obs.columns:\n",
    "            raise KeyError(f\"{groupby} is not a valid obs annotation.\")\n",
    "        groups = adata.obs[groupby].unique()\n",
    "        label_df = adata.obs[[groupby]].astype(str).rename(columns={groupby: key_added})\n",
    "        for grp in groups:\n",
    "            k = adata.obs[groupby] == grp\n",
    "            grp_size = sum(k)\n",
    "            n_partition = max(np.round(grp_size / partition_size).astype(int), 1)\n",
    "            if method == \"random\":\n",
    "                part_idx = np.random.randint(low=0, high=n_partition, size=grp_size)\n",
    "            elif method == \"random_even\":\n",
    "                part_sizes = list(map(len, np.array_split(np.arange(grp_size), n_partition)))\n",
    "                part_idx = np.repeat(np.arange(n_partition), part_sizes)\n",
    "                np.random.shuffle(part_idx)\n",
    "            else:\n",
    "                raise NotImplementedError(method)\n",
    "            label_df.loc[k, key_added] = [f\"{grp},{i}\" for i in part_idx]\n",
    "        adata.obs[key_added] = label_df[key_added]\n",
    "    else:\n",
    "        n_partition = max(np.round(adata.n_obs / partition_size).astype(int), 1)\n",
    "        if method == \"random\":\n",
    "            part_idx = np.random.randint(low=0, high=n_partition, size=adata.n_obs)\n",
    "        elif method == \"random_even\":\n",
    "            part_sizes = list(map(len, np.array_split(np.arange(adata.n_obs), n_partition)))\n",
    "            part_idx = np.repeat(np.arange(n_partition), part_sizes)\n",
    "            np.random.shuffle(part_idx)\n",
    "        else:\n",
    "            raise NotImplementedError(method)\n",
    "        adata.obs[key_added] = part_idx.astype(str)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "satellite-celebration",
   "metadata": {},
   "source": [
    "## Import suspension data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "thick-ancient",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 176050 × 3311\n",
       "    obs: 'latent_RT_efficiency', 'latent_cell_probability', 'latent_scale', 'sangerID', 'combinedID', 'donor', 'donor_type', 'region_finest', 'facility', 'cell_or_nuclei', 'modality', 'kit_10x', 'flushed', 'scrublet_score', 'scrublet_leiden', 'cluster_scrublet_score', 'doublet_pval', 'doublet_bh_pval', 'n_genes', 'n_counts', 'n_genes_by_counts', 'log1p_n_genes_by_counts', 'total_counts', 'log1p_total_counts', 'total_counts_mt', 'log1p_total_counts_mt', 'pct_counts_mt', 'total_counts_ribo', 'log1p_total_counts_ribo', 'pct_counts_ribo', 'HB_score', 'batch_key', '_scvi_batch', '_scvi_labels', 'annotation_Semih', 'leiden_scVI', 'cell_type_pre', 'pth', 'ito', 'FACSgate', 'leiden_scVI_0.05', 'leiden_scVI_0.01', 'leiden_scVI_0.1', 'leiden_scVI_0.15', 'leiden_scVI_0.2', 'leiden_scVI_0.25', 'leiden_scVI_0.3', 'leiden_scVI_0.35', 'leiden_scVI_0.4', 'cell_type', 'coarse_grain', 'fine_grain', 'region', 'sex', 'age', 'cycling', 'S_score', 'G2M_score', 'phase', 'stress_score', 'hb1_score'\n",
       "    var: 'gene_name_scRNA-0', 'gene_id', 'highly_variable', 'means', 'dispersions', 'dispersions_norm', 'highly_variable_nbatches', 'highly_variable_intersection'\n",
       "    uns: '_scvi_manager_uuid', '_scvi_uuid', 'age_colors', 'cell_or_nuclei_colors', 'cell_type_colors', 'coarse_grain_colors', 'cycling_colors', 'donor_colors', 'fine_grain_colors', 'hvg', 'kit_10x_colors', 'neighbors', 'phase_colors', 'region_colors', 'sex_colors', 'umap'\n",
       "    obsm: 'X_scVI', 'X_umap', '_scvi_extra_categorical_covs', '_scvi_extra_continuous', '_scvi_extra_continuous_covs', 'latent_gene_encoding'\n",
       "    layers: 'counts'\n",
       "    obsp: 'connectivities', 'distances'"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata = sc.read('/nfs/team205/heart/anndata_objects/Foetal/final_annotation/FetalHearts13Aug.h5ad')\n",
    "adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "pacific-survey",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(176050, 3311)\n",
      "(176050, 36601)\n"
     ]
    }
   ],
   "source": [
    "# Recover log-normed data for all genes to adata.X (allows Compass to calculate reactions across all genes, not just HVG-subsetted!)\n",
    "print(adata.shape)\n",
    "adata=adata.raw.to_adata()\n",
    "print(adata.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "refined-party",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Index(['MIR1302-2HG', 'FAM138A'], dtype='object', name='gene_name_multiome-1')\n",
      "[0.09802545 0.09802545 0.41538525 0.41538525 0.6009918 ]\n"
     ]
    }
   ],
   "source": [
    "# Check var_names are gene symbols, and the data is log-normed\n",
    "print(adata.var.index[:2])\n",
    "print(adata.X.data[:5])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "conventional-necklace",
   "metadata": {},
   "source": [
    "# Prepare data for Compass\n",
    "\"The input gene expression matrix can be either a tab-delimited text file (tsv) or a matrix market format (mtx) containing gene expression estimates (CPM, TPM, or similar scaled units) with one row per gene, one column per sample.\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "stuck-amsterdam",
   "metadata": {},
   "source": [
    "## .mtx method (gene x sampleID, 3 files)\n",
    "This allows you to specify the 'samples', so rather than calculating reactions in all single cells individually (or micro-pooled) we can make our own \"samples\" (pseudobulking them)\n",
    "\n",
    "Need to ensure the samples are representative of the sample. Here we use a new batch key of cell state and the previous batch key  to form the 'samples'\n",
    "This requires 3 files, to inputted thus ```--data expression.mtx genes.tsv sample_names.tsv```\n",
    "\n",
    "Another option would be to use Milo or Metacell to make small samples, below I am going to just use the batch key to make samples within each cell state."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "right-trunk",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "GreatVesselAdventitialFibroblasts        18605\n",
       "MyocardialInterstitialFibroblasts        15073\n",
       "VentricularCardiomyocytesCompact         13740\n",
       "GreatVesselProgenitors                   13175\n",
       "VentricularCardiomyocytesPRRX1pos        11627\n",
       "VentricularCardiomyocytesCycling         11297\n",
       "MacrophagesLYVE1pos                       8864\n",
       "GreatVesselSmoothMuscleCells              7870\n",
       "EndocardialCells                          7418\n",
       "ValveInterstitialCells                    7222\n",
       "VentricularCardiomyocytesTrabeculated     6936\n",
       "AtrialCardiomyocytesLeft                  6304\n",
       "AtrialCardiomyocytesRight                 5452\n",
       "Cardiofibromyocytes                       5278\n",
       "CoronaryEndothelialCellsCapillary         4834\n",
       "CoronaryVesselAdventitialFibroblasts      2683\n",
       "CoronaryPericytes                         2077\n",
       "SatelliteGlialCells                       2008\n",
       "Myofibroblasts                            1826\n",
       "MacrophagesCX3CR1pos                      1609\n",
       "VentricularConductionSystemDistal         1568\n",
       "InnateLymphoidCells                       1498\n",
       "MesothelialEpicardialCells                1469\n",
       "NaturalKillerCells                        1351\n",
       "CoronarySmoothMuscleCells                 1290\n",
       "Monocytes                                 1260\n",
       "ProBCells                                 1252\n",
       "ValvalEndothelialCells                     908\n",
       "CoronaryEndothelialCellsArterial           885\n",
       "ParasympatheticNeurons                     883\n",
       "GreatVesselEndothelialCells                857\n",
       "SinoatrialNodeCardiomyocytes               814\n",
       "FibroblasticEpicardialCells                745\n",
       "VentricularConductionSystemProximal        730\n",
       "BCells                                     724\n",
       "TCells                                     691\n",
       "AtrioventricularNodeCardiomyocytes         689\n",
       "LymphaticEndothelialCells                  676\n",
       "SatelliteGlialCellProgenitors              671\n",
       "SympatheticNeurons                         580\n",
       "MastCells                                  507\n",
       "NeuronProgenitors                          474\n",
       "DendriticCells                             434\n",
       "Platelets                                  428\n",
       "AtrialCardiomyocytesCycling                316\n",
       "CoronaryEndothelialCellsVenous             270\n",
       "ChromaffinCells                            182\n",
       "Name: fine_grain, dtype: int64"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# make random groups of size 20 within each cell state\n",
    "adata.obs['fine_grain'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "adequate-search",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Make random partition using sctk\n",
    "random_partition(adata, partition_size=20, groupby=\"fine_grain\", key_added=\"partition_20\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "behavioral-senate",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>latent_RT_efficiency</th>\n",
       "      <th>latent_cell_probability</th>\n",
       "      <th>latent_scale</th>\n",
       "      <th>sangerID</th>\n",
       "      <th>combinedID</th>\n",
       "      <th>donor</th>\n",
       "      <th>donor_type</th>\n",
       "      <th>region_finest</th>\n",
       "      <th>facility</th>\n",
       "      <th>cell_or_nuclei</th>\n",
       "      <th>...</th>\n",
       "      <th>region</th>\n",
       "      <th>sex</th>\n",
       "      <th>age</th>\n",
       "      <th>cycling</th>\n",
       "      <th>S_score</th>\n",
       "      <th>G2M_score</th>\n",
       "      <th>phase</th>\n",
       "      <th>stress_score</th>\n",
       "      <th>hb1_score</th>\n",
       "      <th>partition_20</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>barcode</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>BHF_F_Hea10402917_TGGCTGGGTTGTGGAG-1</th>\n",
       "      <td>5.218614</td>\n",
       "      <td>0.999939</td>\n",
       "      <td>123665.210938</td>\n",
       "      <td>BHF_F_Hea10402917</td>\n",
       "      <td>na</td>\n",
       "      <td>C86</td>\n",
       "      <td>na</td>\n",
       "      <td>Heart</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>Cell</td>\n",
       "      <td>...</td>\n",
       "      <td>WholeHeart</td>\n",
       "      <td>Male</td>\n",
       "      <td>12W5D</td>\n",
       "      <td>Yes</td>\n",
       "      <td>-0.005998</td>\n",
       "      <td>-0.126205</td>\n",
       "      <td>G1</td>\n",
       "      <td>0.494416</td>\n",
       "      <td>-0.223034</td>\n",
       "      <td>CoronaryEndothelialCellsCapillary,58</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BHF_F_Hea10402917_ATGTGTGTCGAACGGA-1</th>\n",
       "      <td>3.109560</td>\n",
       "      <td>0.999991</td>\n",
       "      <td>10201.502930</td>\n",
       "      <td>BHF_F_Hea10402917</td>\n",
       "      <td>na</td>\n",
       "      <td>C86</td>\n",
       "      <td>na</td>\n",
       "      <td>Heart</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>Cell</td>\n",
       "      <td>...</td>\n",
       "      <td>WholeHeart</td>\n",
       "      <td>Male</td>\n",
       "      <td>12W5D</td>\n",
       "      <td>Yes</td>\n",
       "      <td>0.656214</td>\n",
       "      <td>0.093990</td>\n",
       "      <td>S</td>\n",
       "      <td>1.074951</td>\n",
       "      <td>-0.143935</td>\n",
       "      <td>InnateLymphoidCells,53</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BHF_F_Hea10402917_CAGGTGCAGGCTATCT-1</th>\n",
       "      <td>3.181446</td>\n",
       "      <td>0.999989</td>\n",
       "      <td>10217.305664</td>\n",
       "      <td>BHF_F_Hea10402917</td>\n",
       "      <td>na</td>\n",
       "      <td>C86</td>\n",
       "      <td>na</td>\n",
       "      <td>Heart</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>Cell</td>\n",
       "      <td>...</td>\n",
       "      <td>WholeHeart</td>\n",
       "      <td>Male</td>\n",
       "      <td>12W5D</td>\n",
       "      <td>Yes</td>\n",
       "      <td>0.386814</td>\n",
       "      <td>1.163369</td>\n",
       "      <td>G2M</td>\n",
       "      <td>0.460513</td>\n",
       "      <td>-0.138523</td>\n",
       "      <td>BCells,5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BHF_F_Hea10402917_TACAGTGTCCAGTATG-1</th>\n",
       "      <td>3.101654</td>\n",
       "      <td>0.999993</td>\n",
       "      <td>9392.824219</td>\n",
       "      <td>BHF_F_Hea10402917</td>\n",
       "      <td>na</td>\n",
       "      <td>C86</td>\n",
       "      <td>na</td>\n",
       "      <td>Heart</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>Cell</td>\n",
       "      <td>...</td>\n",
       "      <td>WholeHeart</td>\n",
       "      <td>Male</td>\n",
       "      <td>12W5D</td>\n",
       "      <td>Yes</td>\n",
       "      <td>0.640767</td>\n",
       "      <td>0.132809</td>\n",
       "      <td>S</td>\n",
       "      <td>0.994759</td>\n",
       "      <td>-0.053963</td>\n",
       "      <td>InnateLymphoidCells,59</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BHF_F_Hea10402917_ACCAGTATCCACTCCA-1</th>\n",
       "      <td>3.109215</td>\n",
       "      <td>0.999989</td>\n",
       "      <td>9608.674805</td>\n",
       "      <td>BHF_F_Hea10402917</td>\n",
       "      <td>na</td>\n",
       "      <td>C86</td>\n",
       "      <td>na</td>\n",
       "      <td>Heart</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>Cell</td>\n",
       "      <td>...</td>\n",
       "      <td>WholeHeart</td>\n",
       "      <td>Male</td>\n",
       "      <td>12W5D</td>\n",
       "      <td>No</td>\n",
       "      <td>-0.107722</td>\n",
       "      <td>-0.040233</td>\n",
       "      <td>G1</td>\n",
       "      <td>0.789726</td>\n",
       "      <td>-0.090271</td>\n",
       "      <td>ProBCells,3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BHF_F_Hea11933675_BHF_F_Hea11596628_GCTCATTGTCAGGCCA-1</th>\n",
       "      <td>0.756771</td>\n",
       "      <td>0.978465</td>\n",
       "      <td>3959.675781</td>\n",
       "      <td>BHF_F_Hea11933675</td>\n",
       "      <td>BHF_F_Hea11933675_BHF_F_Hea11596628</td>\n",
       "      <td>C98</td>\n",
       "      <td>na</td>\n",
       "      <td>Heart</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>Nuclei</td>\n",
       "      <td>...</td>\n",
       "      <td>WholeHeart</td>\n",
       "      <td>Male</td>\n",
       "      <td>9W1D</td>\n",
       "      <td>No</td>\n",
       "      <td>-0.118187</td>\n",
       "      <td>-0.089172</td>\n",
       "      <td>G1</td>\n",
       "      <td>-0.454305</td>\n",
       "      <td>0.097133</td>\n",
       "      <td>GreatVesselAdventitialFibroblasts,929</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BHF_F_Hea11933675_BHF_F_Hea11596628_AAGACATAGCATGACT-1</th>\n",
       "      <td>0.756599</td>\n",
       "      <td>0.978873</td>\n",
       "      <td>4021.003662</td>\n",
       "      <td>BHF_F_Hea11933675</td>\n",
       "      <td>BHF_F_Hea11933675_BHF_F_Hea11596628</td>\n",
       "      <td>C98</td>\n",
       "      <td>na</td>\n",
       "      <td>Heart</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>Nuclei</td>\n",
       "      <td>...</td>\n",
       "      <td>WholeHeart</td>\n",
       "      <td>Male</td>\n",
       "      <td>9W1D</td>\n",
       "      <td>Yes</td>\n",
       "      <td>-0.060168</td>\n",
       "      <td>0.490866</td>\n",
       "      <td>G2M</td>\n",
       "      <td>0.222192</td>\n",
       "      <td>0.669477</td>\n",
       "      <td>GreatVesselSmoothMuscleCells,7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BHF_F_Hea11933675_BHF_F_Hea11596628_AGACAATAGGAGTAAT-1</th>\n",
       "      <td>0.744038</td>\n",
       "      <td>0.971963</td>\n",
       "      <td>3929.745605</td>\n",
       "      <td>BHF_F_Hea11933675</td>\n",
       "      <td>BHF_F_Hea11933675_BHF_F_Hea11596628</td>\n",
       "      <td>C98</td>\n",
       "      <td>na</td>\n",
       "      <td>Heart</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>Nuclei</td>\n",
       "      <td>...</td>\n",
       "      <td>WholeHeart</td>\n",
       "      <td>Male</td>\n",
       "      <td>9W1D</td>\n",
       "      <td>No</td>\n",
       "      <td>-0.152945</td>\n",
       "      <td>-0.124303</td>\n",
       "      <td>G1</td>\n",
       "      <td>-0.028035</td>\n",
       "      <td>0.766656</td>\n",
       "      <td>Monocytes,8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BHF_F_Hea11933675_BHF_F_Hea11596628_GGCCTTAAGGCCTAAT-1</th>\n",
       "      <td>0.769857</td>\n",
       "      <td>0.965501</td>\n",
       "      <td>3725.604004</td>\n",
       "      <td>BHF_F_Hea11933675</td>\n",
       "      <td>BHF_F_Hea11933675_BHF_F_Hea11596628</td>\n",
       "      <td>C98</td>\n",
       "      <td>na</td>\n",
       "      <td>Heart</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>Nuclei</td>\n",
       "      <td>...</td>\n",
       "      <td>WholeHeart</td>\n",
       "      <td>Male</td>\n",
       "      <td>9W1D</td>\n",
       "      <td>Yes</td>\n",
       "      <td>0.115479</td>\n",
       "      <td>-0.161365</td>\n",
       "      <td>S</td>\n",
       "      <td>-0.015999</td>\n",
       "      <td>0.371164</td>\n",
       "      <td>Cardiofibromyocytes,191</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BHF_F_Hea11933675_BHF_F_Hea11596628_AACATAGCAGCAAGGC-1</th>\n",
       "      <td>0.754149</td>\n",
       "      <td>0.969744</td>\n",
       "      <td>3803.776123</td>\n",
       "      <td>BHF_F_Hea11933675</td>\n",
       "      <td>BHF_F_Hea11933675_BHF_F_Hea11596628</td>\n",
       "      <td>C98</td>\n",
       "      <td>na</td>\n",
       "      <td>Heart</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>Nuclei</td>\n",
       "      <td>...</td>\n",
       "      <td>WholeHeart</td>\n",
       "      <td>Male</td>\n",
       "      <td>9W1D</td>\n",
       "      <td>Yes</td>\n",
       "      <td>0.351178</td>\n",
       "      <td>0.041675</td>\n",
       "      <td>S</td>\n",
       "      <td>-0.377977</td>\n",
       "      <td>-0.176750</td>\n",
       "      <td>ValveInterstitialCells,177</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>176050 rows × 62 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                    latent_RT_efficiency  \\\n",
       "barcode                                                                    \n",
       "BHF_F_Hea10402917_TGGCTGGGTTGTGGAG-1                            5.218614   \n",
       "BHF_F_Hea10402917_ATGTGTGTCGAACGGA-1                            3.109560   \n",
       "BHF_F_Hea10402917_CAGGTGCAGGCTATCT-1                            3.181446   \n",
       "BHF_F_Hea10402917_TACAGTGTCCAGTATG-1                            3.101654   \n",
       "BHF_F_Hea10402917_ACCAGTATCCACTCCA-1                            3.109215   \n",
       "...                                                                  ...   \n",
       "BHF_F_Hea11933675_BHF_F_Hea11596628_GCTCATTGTCA...              0.756771   \n",
       "BHF_F_Hea11933675_BHF_F_Hea11596628_AAGACATAGCA...              0.756599   \n",
       "BHF_F_Hea11933675_BHF_F_Hea11596628_AGACAATAGGA...              0.744038   \n",
       "BHF_F_Hea11933675_BHF_F_Hea11596628_GGCCTTAAGGC...              0.769857   \n",
       "BHF_F_Hea11933675_BHF_F_Hea11596628_AACATAGCAGC...              0.754149   \n",
       "\n",
       "                                                    latent_cell_probability  \\\n",
       "barcode                                                                       \n",
       "BHF_F_Hea10402917_TGGCTGGGTTGTGGAG-1                               0.999939   \n",
       "BHF_F_Hea10402917_ATGTGTGTCGAACGGA-1                               0.999991   \n",
       "BHF_F_Hea10402917_CAGGTGCAGGCTATCT-1                               0.999989   \n",
       "BHF_F_Hea10402917_TACAGTGTCCAGTATG-1                               0.999993   \n",
       "BHF_F_Hea10402917_ACCAGTATCCACTCCA-1                               0.999989   \n",
       "...                                                                     ...   \n",
       "BHF_F_Hea11933675_BHF_F_Hea11596628_GCTCATTGTCA...                 0.978465   \n",
       "BHF_F_Hea11933675_BHF_F_Hea11596628_AAGACATAGCA...                 0.978873   \n",
       "BHF_F_Hea11933675_BHF_F_Hea11596628_AGACAATAGGA...                 0.971963   \n",
       "BHF_F_Hea11933675_BHF_F_Hea11596628_GGCCTTAAGGC...                 0.965501   \n",
       "BHF_F_Hea11933675_BHF_F_Hea11596628_AACATAGCAGC...                 0.969744   \n",
       "\n",
       "                                                     latent_scale  \\\n",
       "barcode                                                             \n",
       "BHF_F_Hea10402917_TGGCTGGGTTGTGGAG-1                123665.210938   \n",
       "BHF_F_Hea10402917_ATGTGTGTCGAACGGA-1                 10201.502930   \n",
       "BHF_F_Hea10402917_CAGGTGCAGGCTATCT-1                 10217.305664   \n",
       "BHF_F_Hea10402917_TACAGTGTCCAGTATG-1                  9392.824219   \n",
       "BHF_F_Hea10402917_ACCAGTATCCACTCCA-1                  9608.674805   \n",
       "...                                                           ...   \n",
       "BHF_F_Hea11933675_BHF_F_Hea11596628_GCTCATTGTCA...    3959.675781   \n",
       "BHF_F_Hea11933675_BHF_F_Hea11596628_AAGACATAGCA...    4021.003662   \n",
       "BHF_F_Hea11933675_BHF_F_Hea11596628_AGACAATAGGA...    3929.745605   \n",
       "BHF_F_Hea11933675_BHF_F_Hea11596628_GGCCTTAAGGC...    3725.604004   \n",
       "BHF_F_Hea11933675_BHF_F_Hea11596628_AACATAGCAGC...    3803.776123   \n",
       "\n",
       "                                                             sangerID  \\\n",
       "barcode                                                                 \n",
       "BHF_F_Hea10402917_TGGCTGGGTTGTGGAG-1                BHF_F_Hea10402917   \n",
       "BHF_F_Hea10402917_ATGTGTGTCGAACGGA-1                BHF_F_Hea10402917   \n",
       "BHF_F_Hea10402917_CAGGTGCAGGCTATCT-1                BHF_F_Hea10402917   \n",
       "BHF_F_Hea10402917_TACAGTGTCCAGTATG-1                BHF_F_Hea10402917   \n",
       "BHF_F_Hea10402917_ACCAGTATCCACTCCA-1                BHF_F_Hea10402917   \n",
       "...                                                               ...   \n",
       "BHF_F_Hea11933675_BHF_F_Hea11596628_GCTCATTGTCA...  BHF_F_Hea11933675   \n",
       "BHF_F_Hea11933675_BHF_F_Hea11596628_AAGACATAGCA...  BHF_F_Hea11933675   \n",
       "BHF_F_Hea11933675_BHF_F_Hea11596628_AGACAATAGGA...  BHF_F_Hea11933675   \n",
       "BHF_F_Hea11933675_BHF_F_Hea11596628_GGCCTTAAGGC...  BHF_F_Hea11933675   \n",
       "BHF_F_Hea11933675_BHF_F_Hea11596628_AACATAGCAGC...  BHF_F_Hea11933675   \n",
       "\n",
       "                                                                             combinedID  \\\n",
       "barcode                                                                                   \n",
       "BHF_F_Hea10402917_TGGCTGGGTTGTGGAG-1                                                 na   \n",
       "BHF_F_Hea10402917_ATGTGTGTCGAACGGA-1                                                 na   \n",
       "BHF_F_Hea10402917_CAGGTGCAGGCTATCT-1                                                 na   \n",
       "BHF_F_Hea10402917_TACAGTGTCCAGTATG-1                                                 na   \n",
       "BHF_F_Hea10402917_ACCAGTATCCACTCCA-1                                                 na   \n",
       "...                                                                                 ...   \n",
       "BHF_F_Hea11933675_BHF_F_Hea11596628_GCTCATTGTCA...  BHF_F_Hea11933675_BHF_F_Hea11596628   \n",
       "BHF_F_Hea11933675_BHF_F_Hea11596628_AAGACATAGCA...  BHF_F_Hea11933675_BHF_F_Hea11596628   \n",
       "BHF_F_Hea11933675_BHF_F_Hea11596628_AGACAATAGGA...  BHF_F_Hea11933675_BHF_F_Hea11596628   \n",
       "BHF_F_Hea11933675_BHF_F_Hea11596628_GGCCTTAAGGC...  BHF_F_Hea11933675_BHF_F_Hea11596628   \n",
       "BHF_F_Hea11933675_BHF_F_Hea11596628_AACATAGCAGC...  BHF_F_Hea11933675_BHF_F_Hea11596628   \n",
       "\n",
       "                                                   donor donor_type  \\\n",
       "barcode                                                               \n",
       "BHF_F_Hea10402917_TGGCTGGGTTGTGGAG-1                 C86         na   \n",
       "BHF_F_Hea10402917_ATGTGTGTCGAACGGA-1                 C86         na   \n",
       "BHF_F_Hea10402917_CAGGTGCAGGCTATCT-1                 C86         na   \n",
       "BHF_F_Hea10402917_TACAGTGTCCAGTATG-1                 C86         na   \n",
       "BHF_F_Hea10402917_ACCAGTATCCACTCCA-1                 C86         na   \n",
       "...                                                  ...        ...   \n",
       "BHF_F_Hea11933675_BHF_F_Hea11596628_GCTCATTGTCA...   C98         na   \n",
       "BHF_F_Hea11933675_BHF_F_Hea11596628_AAGACATAGCA...   C98         na   \n",
       "BHF_F_Hea11933675_BHF_F_Hea11596628_AGACAATAGGA...   C98         na   \n",
       "BHF_F_Hea11933675_BHF_F_Hea11596628_GGCCTTAAGGC...   C98         na   \n",
       "BHF_F_Hea11933675_BHF_F_Hea11596628_AACATAGCAGC...   C98         na   \n",
       "\n",
       "                                                   region_finest facility  \\\n",
       "barcode                                                                     \n",
       "BHF_F_Hea10402917_TGGCTGGGTTGTGGAG-1                       Heart   Sanger   \n",
       "BHF_F_Hea10402917_ATGTGTGTCGAACGGA-1                       Heart   Sanger   \n",
       "BHF_F_Hea10402917_CAGGTGCAGGCTATCT-1                       Heart   Sanger   \n",
       "BHF_F_Hea10402917_TACAGTGTCCAGTATG-1                       Heart   Sanger   \n",
       "BHF_F_Hea10402917_ACCAGTATCCACTCCA-1                       Heart   Sanger   \n",
       "...                                                          ...      ...   \n",
       "BHF_F_Hea11933675_BHF_F_Hea11596628_GCTCATTGTCA...         Heart   Sanger   \n",
       "BHF_F_Hea11933675_BHF_F_Hea11596628_AAGACATAGCA...         Heart   Sanger   \n",
       "BHF_F_Hea11933675_BHF_F_Hea11596628_AGACAATAGGA...         Heart   Sanger   \n",
       "BHF_F_Hea11933675_BHF_F_Hea11596628_GGCCTTAAGGC...         Heart   Sanger   \n",
       "BHF_F_Hea11933675_BHF_F_Hea11596628_AACATAGCAGC...         Heart   Sanger   \n",
       "\n",
       "                                                   cell_or_nuclei  ...  \\\n",
       "barcode                                                            ...   \n",
       "BHF_F_Hea10402917_TGGCTGGGTTGTGGAG-1                         Cell  ...   \n",
       "BHF_F_Hea10402917_ATGTGTGTCGAACGGA-1                         Cell  ...   \n",
       "BHF_F_Hea10402917_CAGGTGCAGGCTATCT-1                         Cell  ...   \n",
       "BHF_F_Hea10402917_TACAGTGTCCAGTATG-1                         Cell  ...   \n",
       "BHF_F_Hea10402917_ACCAGTATCCACTCCA-1                         Cell  ...   \n",
       "...                                                           ...  ...   \n",
       "BHF_F_Hea11933675_BHF_F_Hea11596628_GCTCATTGTCA...         Nuclei  ...   \n",
       "BHF_F_Hea11933675_BHF_F_Hea11596628_AAGACATAGCA...         Nuclei  ...   \n",
       "BHF_F_Hea11933675_BHF_F_Hea11596628_AGACAATAGGA...         Nuclei  ...   \n",
       "BHF_F_Hea11933675_BHF_F_Hea11596628_GGCCTTAAGGC...         Nuclei  ...   \n",
       "BHF_F_Hea11933675_BHF_F_Hea11596628_AACATAGCAGC...         Nuclei  ...   \n",
       "\n",
       "                                                        region   sex    age  \\\n",
       "barcode                                                                       \n",
       "BHF_F_Hea10402917_TGGCTGGGTTGTGGAG-1                WholeHeart  Male  12W5D   \n",
       "BHF_F_Hea10402917_ATGTGTGTCGAACGGA-1                WholeHeart  Male  12W5D   \n",
       "BHF_F_Hea10402917_CAGGTGCAGGCTATCT-1                WholeHeart  Male  12W5D   \n",
       "BHF_F_Hea10402917_TACAGTGTCCAGTATG-1                WholeHeart  Male  12W5D   \n",
       "BHF_F_Hea10402917_ACCAGTATCCACTCCA-1                WholeHeart  Male  12W5D   \n",
       "...                                                        ...   ...    ...   \n",
       "BHF_F_Hea11933675_BHF_F_Hea11596628_GCTCATTGTCA...  WholeHeart  Male   9W1D   \n",
       "BHF_F_Hea11933675_BHF_F_Hea11596628_AAGACATAGCA...  WholeHeart  Male   9W1D   \n",
       "BHF_F_Hea11933675_BHF_F_Hea11596628_AGACAATAGGA...  WholeHeart  Male   9W1D   \n",
       "BHF_F_Hea11933675_BHF_F_Hea11596628_GGCCTTAAGGC...  WholeHeart  Male   9W1D   \n",
       "BHF_F_Hea11933675_BHF_F_Hea11596628_AACATAGCAGC...  WholeHeart  Male   9W1D   \n",
       "\n",
       "                                                    cycling   S_score  \\\n",
       "barcode                                                                 \n",
       "BHF_F_Hea10402917_TGGCTGGGTTGTGGAG-1                    Yes -0.005998   \n",
       "BHF_F_Hea10402917_ATGTGTGTCGAACGGA-1                    Yes  0.656214   \n",
       "BHF_F_Hea10402917_CAGGTGCAGGCTATCT-1                    Yes  0.386814   \n",
       "BHF_F_Hea10402917_TACAGTGTCCAGTATG-1                    Yes  0.640767   \n",
       "BHF_F_Hea10402917_ACCAGTATCCACTCCA-1                     No -0.107722   \n",
       "...                                                     ...       ...   \n",
       "BHF_F_Hea11933675_BHF_F_Hea11596628_GCTCATTGTCA...       No -0.118187   \n",
       "BHF_F_Hea11933675_BHF_F_Hea11596628_AAGACATAGCA...      Yes -0.060168   \n",
       "BHF_F_Hea11933675_BHF_F_Hea11596628_AGACAATAGGA...       No -0.152945   \n",
       "BHF_F_Hea11933675_BHF_F_Hea11596628_GGCCTTAAGGC...      Yes  0.115479   \n",
       "BHF_F_Hea11933675_BHF_F_Hea11596628_AACATAGCAGC...      Yes  0.351178   \n",
       "\n",
       "                                                    G2M_score  phase  \\\n",
       "barcode                                                                \n",
       "BHF_F_Hea10402917_TGGCTGGGTTGTGGAG-1                -0.126205     G1   \n",
       "BHF_F_Hea10402917_ATGTGTGTCGAACGGA-1                 0.093990      S   \n",
       "BHF_F_Hea10402917_CAGGTGCAGGCTATCT-1                 1.163369    G2M   \n",
       "BHF_F_Hea10402917_TACAGTGTCCAGTATG-1                 0.132809      S   \n",
       "BHF_F_Hea10402917_ACCAGTATCCACTCCA-1                -0.040233     G1   \n",
       "...                                                       ...    ...   \n",
       "BHF_F_Hea11933675_BHF_F_Hea11596628_GCTCATTGTCA...  -0.089172     G1   \n",
       "BHF_F_Hea11933675_BHF_F_Hea11596628_AAGACATAGCA...   0.490866    G2M   \n",
       "BHF_F_Hea11933675_BHF_F_Hea11596628_AGACAATAGGA...  -0.124303     G1   \n",
       "BHF_F_Hea11933675_BHF_F_Hea11596628_GGCCTTAAGGC...  -0.161365      S   \n",
       "BHF_F_Hea11933675_BHF_F_Hea11596628_AACATAGCAGC...   0.041675      S   \n",
       "\n",
       "                                                    stress_score  hb1_score  \\\n",
       "barcode                                                                       \n",
       "BHF_F_Hea10402917_TGGCTGGGTTGTGGAG-1                    0.494416  -0.223034   \n",
       "BHF_F_Hea10402917_ATGTGTGTCGAACGGA-1                    1.074951  -0.143935   \n",
       "BHF_F_Hea10402917_CAGGTGCAGGCTATCT-1                    0.460513  -0.138523   \n",
       "BHF_F_Hea10402917_TACAGTGTCCAGTATG-1                    0.994759  -0.053963   \n",
       "BHF_F_Hea10402917_ACCAGTATCCACTCCA-1                    0.789726  -0.090271   \n",
       "...                                                          ...        ...   \n",
       "BHF_F_Hea11933675_BHF_F_Hea11596628_GCTCATTGTCA...     -0.454305   0.097133   \n",
       "BHF_F_Hea11933675_BHF_F_Hea11596628_AAGACATAGCA...      0.222192   0.669477   \n",
       "BHF_F_Hea11933675_BHF_F_Hea11596628_AGACAATAGGA...     -0.028035   0.766656   \n",
       "BHF_F_Hea11933675_BHF_F_Hea11596628_GGCCTTAAGGC...     -0.015999   0.371164   \n",
       "BHF_F_Hea11933675_BHF_F_Hea11596628_AACATAGCAGC...     -0.377977  -0.176750   \n",
       "\n",
       "                                                                             partition_20  \n",
       "barcode                                                                                    \n",
       "BHF_F_Hea10402917_TGGCTGGGTTGTGGAG-1                 CoronaryEndothelialCellsCapillary,58  \n",
       "BHF_F_Hea10402917_ATGTGTGTCGAACGGA-1                               InnateLymphoidCells,53  \n",
       "BHF_F_Hea10402917_CAGGTGCAGGCTATCT-1                                             BCells,5  \n",
       "BHF_F_Hea10402917_TACAGTGTCCAGTATG-1                               InnateLymphoidCells,59  \n",
       "BHF_F_Hea10402917_ACCAGTATCCACTCCA-1                                          ProBCells,3  \n",
       "...                                                                                   ...  \n",
       "BHF_F_Hea11933675_BHF_F_Hea11596628_GCTCATTGTCA...  GreatVesselAdventitialFibroblasts,929  \n",
       "BHF_F_Hea11933675_BHF_F_Hea11596628_AAGACATAGCA...         GreatVesselSmoothMuscleCells,7  \n",
       "BHF_F_Hea11933675_BHF_F_Hea11596628_AGACAATAGGA...                            Monocytes,8  \n",
       "BHF_F_Hea11933675_BHF_F_Hea11596628_GGCCTTAAGGC...                Cardiofibromyocytes,191  \n",
       "BHF_F_Hea11933675_BHF_F_Hea11596628_AACATAGCAGC...             ValveInterstitialCells,177  \n",
       "\n",
       "[176050 rows x 62 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.obs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "advanced-there",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Pseudobulk use Ni's psudobulk function\n",
    "pseudo_bulk = pseudo_bulk(adata,\n",
    "                         groupby='partition_20')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "antique-there",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(36601, 8802)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>AtrialCardiomyocytesCycling,0</th>\n",
       "      <th>AtrialCardiomyocytesCycling,1</th>\n",
       "      <th>AtrialCardiomyocytesCycling,10</th>\n",
       "      <th>AtrialCardiomyocytesCycling,11</th>\n",
       "      <th>AtrialCardiomyocytesCycling,12</th>\n",
       "      <th>AtrialCardiomyocytesCycling,13</th>\n",
       "      <th>AtrialCardiomyocytesCycling,14</th>\n",
       "      <th>AtrialCardiomyocytesCycling,15</th>\n",
       "      <th>AtrialCardiomyocytesCycling,2</th>\n",
       "      <th>AtrialCardiomyocytesCycling,3</th>\n",
       "      <th>...</th>\n",
       "      <th>VentricularConductionSystemProximal,32</th>\n",
       "      <th>VentricularConductionSystemProximal,33</th>\n",
       "      <th>VentricularConductionSystemProximal,34</th>\n",
       "      <th>VentricularConductionSystemProximal,35</th>\n",
       "      <th>VentricularConductionSystemProximal,4</th>\n",
       "      <th>VentricularConductionSystemProximal,5</th>\n",
       "      <th>VentricularConductionSystemProximal,6</th>\n",
       "      <th>VentricularConductionSystemProximal,7</th>\n",
       "      <th>VentricularConductionSystemProximal,8</th>\n",
       "      <th>VentricularConductionSystemProximal,9</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>MIR1302-2HG</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FAM138A</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OR4F5</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 8802 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             AtrialCardiomyocytesCycling,0  AtrialCardiomyocytesCycling,1  \\\n",
       "MIR1302-2HG                            0.0                            0.0   \n",
       "FAM138A                                0.0                            0.0   \n",
       "OR4F5                                  0.0                            0.0   \n",
       "\n",
       "             AtrialCardiomyocytesCycling,10  AtrialCardiomyocytesCycling,11  \\\n",
       "MIR1302-2HG                             0.0                             0.0   \n",
       "FAM138A                                 0.0                             0.0   \n",
       "OR4F5                                   0.0                             0.0   \n",
       "\n",
       "             AtrialCardiomyocytesCycling,12  AtrialCardiomyocytesCycling,13  \\\n",
       "MIR1302-2HG                             0.0                             0.0   \n",
       "FAM138A                                 0.0                             0.0   \n",
       "OR4F5                                   0.0                             0.0   \n",
       "\n",
       "             AtrialCardiomyocytesCycling,14  AtrialCardiomyocytesCycling,15  \\\n",
       "MIR1302-2HG                             0.0                             0.0   \n",
       "FAM138A                                 0.0                             0.0   \n",
       "OR4F5                                   0.0                             0.0   \n",
       "\n",
       "             AtrialCardiomyocytesCycling,2  AtrialCardiomyocytesCycling,3  \\\n",
       "MIR1302-2HG                            0.0                            0.0   \n",
       "FAM138A                                0.0                            0.0   \n",
       "OR4F5                                  0.0                            0.0   \n",
       "\n",
       "             ...  VentricularConductionSystemProximal,32  \\\n",
       "MIR1302-2HG  ...                                     0.0   \n",
       "FAM138A      ...                                     0.0   \n",
       "OR4F5        ...                                     0.0   \n",
       "\n",
       "             VentricularConductionSystemProximal,33  \\\n",
       "MIR1302-2HG                                     0.0   \n",
       "FAM138A                                         0.0   \n",
       "OR4F5                                           0.0   \n",
       "\n",
       "             VentricularConductionSystemProximal,34  \\\n",
       "MIR1302-2HG                                     0.0   \n",
       "FAM138A                                         0.0   \n",
       "OR4F5                                           0.0   \n",
       "\n",
       "             VentricularConductionSystemProximal,35  \\\n",
       "MIR1302-2HG                                     0.0   \n",
       "FAM138A                                         0.0   \n",
       "OR4F5                                           0.0   \n",
       "\n",
       "             VentricularConductionSystemProximal,4  \\\n",
       "MIR1302-2HG                                    0.0   \n",
       "FAM138A                                        0.0   \n",
       "OR4F5                                          0.0   \n",
       "\n",
       "             VentricularConductionSystemProximal,5  \\\n",
       "MIR1302-2HG                                    0.0   \n",
       "FAM138A                                        0.0   \n",
       "OR4F5                                          0.0   \n",
       "\n",
       "             VentricularConductionSystemProximal,6  \\\n",
       "MIR1302-2HG                                    0.0   \n",
       "FAM138A                                        0.0   \n",
       "OR4F5                                          0.0   \n",
       "\n",
       "             VentricularConductionSystemProximal,7  \\\n",
       "MIR1302-2HG                                    0.0   \n",
       "FAM138A                                        0.0   \n",
       "OR4F5                                          0.0   \n",
       "\n",
       "             VentricularConductionSystemProximal,8  \\\n",
       "MIR1302-2HG                                    0.0   \n",
       "FAM138A                                        0.0   \n",
       "OR4F5                                          0.0   \n",
       "\n",
       "             VentricularConductionSystemProximal,9  \n",
       "MIR1302-2HG                                    0.0  \n",
       "FAM138A                                        0.0  \n",
       "OR4F5                                          0.0  \n",
       "\n",
       "[3 rows x 8802 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(pseudo_bulk.shape)\n",
    "pseudo_bulk.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "medieval-tiger",
   "metadata": {},
   "outputs": [],
   "source": [
    "# make sample_names.tsv\n",
    "pseudo_bulk.columns.to_frame().to_csv('/nfs/team205/heart/compass/fetal_allcells_random20/sample_names.tsv', sep=\"\\t\", index=False, header=False)\n",
    "\n",
    "# make genes.tsv\n",
    "genes = pseudo_bulk.index.to_list()\n",
    "genes = pd.DataFrame(genes)\n",
    "genes.to_csv('/nfs/team205/heart/compass/fetal_allcells_random20/genes.tsv', sep=\"\\t\", index = False, header=False)\n",
    "\n",
    "# make expression.mtx\n",
    "from scipy.io import mmwrite, mmread\n",
    "from scipy.sparse import csr_matrix\n",
    "\n",
    "# Count Matrix (log normed, all genes, not HVG subsetted)\n",
    "sparce_mtx = csr_matrix(pseudo_bulk)\n",
    "sparce_mtx = sparce_mtx\n",
    "mmwrite('/nfs/team205/heart/compass/fetal_allcells_random20/expression.mtx', sparce_mtx)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "banner-glucose",
   "metadata": {},
   "source": [
    "# Export cell metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "injured-appendix",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Alter so it includes things you need (a reference e.g. barcode and a cell type e.g. in fine_grain)\n",
    "adata.obs.to_csv('/nfs/team205/heart/compass/fetal_allcells_random20/cell_metadata.csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fifteen-remove",
   "metadata": {},
   "source": [
    "# Run Compass"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "confirmed-liability",
   "metadata": {},
   "source": [
    "## bsub command\n",
    "\n",
    "```\n",
    "bsub \\\n",
    "-G teichlab \\\n",
    "-q basement \\\n",
    "-M300000 \\\n",
    "-R \"select[mem>300000] rusage[mem=300000] span[hosts=1]\" \\\n",
    "-n 30 \\\n",
    "-J all_pool50 \\\n",
    "-o 'log/compass.%J.out' \\\n",
    "-e 'log/compass.%J.err' \\\n",
    "'/software/singularity-v3.6.4/bin/singularity run -B /nfs,/lustre -B ~/COMPASSResources:/usr/local/lib/python3.8/site-packages/compass/Resources /nfs/cellgeni/singularity/images/compass-v0.9.10.2-cplex-v20.10.sif compass --data /nfs/team205/heart/compass/fetal_allcells/fetal_allcells_count_matrix.tsv --species homo_sapiens --precache --calc-metabolites --output-dir /nfs/team205/heart/compass/fetal_allcells/out/ --detailed-perf --num-processes 30 --microcluster-size 50'\n",
    "```\n",
    "\n",
    "remove --microcluster-size flag to run on single-cell data\n",
    "\n",
    "if using .mtx methods then need to change --data input e.g. \"--data expression.mtx genes.tsv sample_names.tsv\""
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "scanpy",
   "language": "python",
   "name": "scanpy"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
